{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load libraries \n",
    "import pandas as pd\n",
    "import sspa\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('../Data/Su_COVID_metabolomics_processed_commoncases.csv', index_col=0)\n",
    "reactome_pathways = sspa.process_gmt(\"../Data/Reactome_Homo_sapiens_pathways_compounds_R84.gmt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('../Data/Su_COVID_proteomics_processed_commoncases.csv', index_col=0)\n",
    "reactome_pathways = sspa.process_reactome('Homo sapiens', infile = '../Data/UniProt2Reactome_all_Levels_ver84.txt', download_latest = False, filepath = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Convert pathway ID to name\n",
    "root_path = pd.read_excel('../Data/Root_pathways.xlsx', header=None)\n",
    "root_pathway_dict = {root_path[0][i]:root_path[1][i] for i in range(0,len(root_path))}\n",
    "\n",
    "root_pathway_names = list(root_pathway_dict.keys())\n",
    "#Remove root pathways\n",
    "reactome_pathways = reactome_pathways[~reactome_pathways.index.isin(root_pathway_names)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\jenna\\AppData\\Local\\Temp\\ipykernel_18012\\897059123.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['Group'][i] = 'Severe'\n",
      "C:\\Users\\jenna\\AppData\\Local\\Temp\\ipykernel_18012\\897059123.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['Group'][i] = 'Mild'\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(df.index)):\n",
    "    if df.WHO_status[i] == '1-2':\n",
    "        df['Group'][i] = 'Mild'\n",
    "    else:\n",
    "        df['Group'][i] = 'Severe'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Over-representation analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Initiate an ORA object \n",
    "ora = sspa.sspa_ora(df.iloc[:,:-2], df[\"Group\"], reactome_pathways, 0.05, custom_background=None)\n",
    "#Carry out ORA\n",
    "ora_res = ora.over_representation_analysis()\n",
    "print(len(ora.DA_molecules))\n",
    "display(ora.DA_test_res.sort_values(by=\"P-value\"))\n",
    "display(ora_res.sort_values(by=\"P-value\"))\n",
    "\n",
    "\n",
    "top_10_pathways = ora_res.sort_values(by=\"P-value\").iloc[0:10, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Change name to have a line break otherwise it won't fit on the plot\n",
    "top_10_pathways.Pathway_name[62] = '\\n Transport of inorganic cations/anions and \\n amino acids/oligopeptides   '"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plt.figure(figsize=(9, 5))\n",
    "\n",
    "sns.set(font_scale = 2)\n",
    "sns.set_style(\"ticks\") # same as \"white\" but with ticks\n",
    "\n",
    "#If you want to colour by significance\n",
    "#bar_color = ['tab:green' if float(i) < 0.05 else 'tab:grey' for i in top_20_pathways['P-value']]\n",
    "#sns.barplot(data=top_10_pathways, y=\"Pathway_name\", x=\"P-value\", orient=\"h\", palette=bar_color) #rocket, magma\n",
    "\n",
    "#If you want to colour by gradient\n",
    "ax = sns.barplot(data=top_10_pathways, y=\"Pathway_name\", x=\"P-value\", orient=\"h\", palette=\"rocket\",dodge=False, width=0.8) #rocket, magma\n",
    "\n",
    "#Increase space between the bars and the axis\n",
    "ax.relim()\n",
    "ax.autoscale_view()\n",
    "\n",
    "\n",
    "#Add pathway coverage\n",
    "label = []\n",
    "for i in top_10_pathways.Coverage:\n",
    "    num = i.split('/')\n",
    "    print(num)\n",
    "    percent = (int(num[0]) / int(num[1])) *100\n",
    "    label.append(\"  \"+ str(round(percent,1))+ '%')\n",
    "\n",
    "print(label)\n",
    "\n",
    "ax.bar_label(ax.containers[0], labels=label)\n",
    "\n",
    "#plt.title('Metabolomics',fontsize=22,pad=10)\n",
    "plt.xlabel('Unadjusted p-value',fontsize=26)\n",
    "plt.ylabel('Pathway name',fontsize=26) \n",
    "plt.xlim(0, 0.12)\n",
    "\n",
    "\n",
    "plt.axvline(0.03, c=\"black\")\n",
    "\n",
    "\n",
    "#plt.savefig( '../Figures/proteomic_ORA_top_10.png' , dpi=300,bbox_inches = 'tight' , pad_inches = 0.2 , facecolor='w')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list(top_10_pathways.Coverage)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For integrated data, take the two ORA results and combine the p-values with Fisher's method, which is what is commonly done (See Maghsoudi et al., 2021 for examples):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabolomic_ora = ora_res.sort_values(by=\"P-value\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteomic_ora = ora_res.sort_values(by=\"P-value\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabolomic_ora.index = metabolomic_ora[\"ID\"]\n",
    "metabolomic_ora.drop(columns = [\"ID\"]) #using Sara's code to drop root pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteomic_ora.index = proteomic_ora[\"ID\"]\n",
    "proteomic_ora.drop(columns = [\"ID\"]) #using Sara's code to drop root pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result = metabolomic_ora.merge(proteomic_ora, how='inner',right_index=True, left_index = True)\n",
    "result[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result[\"combined_pval\"] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(len(result.index)):\n",
    "    ID_list = [result[\"P-value_x\"][i],result[\"P-value_y\"][i]]\n",
    "    print(ID_list)\n",
    "    test_stat,pval = scipy.stats.combine_pvalues(ID_list, method='fisher',weights=None)\n",
    "    print(pval)\n",
    "    result[\"combined_pval\"][i] = pval\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result.Coverage_x[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_10_pathways = result.sort_values(by=\"combined_pval\").iloc[0:10, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Obtain pathway coverage for the top 10 results\n",
    "df = pd.read_csv(\"../Data/Su_integrated_data.csv\", index_col=0)\n",
    "reactome_pathways = pd.read_csv(\"../Data/Reactome_multi_omics_ChEBI_Uniprot.csv\", index_col=0,dtype=\"str\") #Dtype warning because in some columns, some values are in string format whereas some are in integer format, that's why I specify dtype=\"str\"\n",
    "kpca_scores = sspa.sspa_kpca(df.iloc[:,:-2], reactome_pathways)\n",
    "\n",
    "\n",
    "#Convert pathway ID to name\n",
    "root_path = pd.read_excel('../Data/Root_pathways.xlsx', header=None)\n",
    "root_pathway_dict = {root_path[0][i]:root_path[1][i] for i in range(0,len(root_path))}\n",
    "\n",
    "root_pathway_names = list(root_pathway_dict.keys())\n",
    "#Using Sara's code, remove root pathways\n",
    "kpca_scores = kpca_scores.drop(columns = list(set(root_pathway_names) & set(kpca_scores.columns)))\n",
    "\n",
    "\n",
    "\n",
    "#Filter out the molecules in the pathways that are not present in the dataset\n",
    "#Obtain all unique values in dataset\n",
    "compounds_present = list(df.columns[:-2])\n",
    "filtered_dict = {} \n",
    "\n",
    "#Obtain pathways and corresponding molecules for all Reactome pathways, store as dictionary\n",
    "orig_dict = sspa.utils.pathwaydf_to_dict(reactome_pathways)\n",
    "\n",
    "#Filter out dictionary to retain only the pathways that remain after kPCA\n",
    "my_keys = kpca_scores.columns\n",
    "pathways_dict = {key: orig_dict[key] for key in my_keys}\n",
    "\n",
    "\n",
    "#My code adapted from Cecilia's\n",
    "#If the key values are not part of the compounds in dataset then remove\n",
    "for key,value in pathways_dict.items():\n",
    "    new_val = [item for item in value if item in compounds_present]\n",
    "    if len(new_val) >= 2: #at least two compounds in the pathway\n",
    "        filtered_dict[key] = new_val"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plt.figure(figsize=(9, 5.5))\n",
    "sns.set(font_scale = 2)\n",
    "\n",
    "sns.set_style(\"ticks\") # same as \"white\" but with ticks\n",
    "\n",
    "ax = sns.barplot(data=top_10_pathways, y=\"Pathway_name_x\", x=\"combined_pval\", orient=\"h\", palette=\"rocket\") #rocket, magma\n",
    "\n",
    "\n",
    "#Increase space between the bars and the axis\n",
    "ax.relim()\n",
    "ax.autoscale_view()\n",
    "\n",
    "\n",
    "\n",
    "label = []\n",
    "# percent1 = []\n",
    "# percent2 = []\n",
    "\n",
    "#Getting pathway coverage by taking the average of the metabolomic and proteomic pathway coverage (not ideal)\n",
    "# for i in range(10):\n",
    "#     pathway1 =  top_10_pathways.Coverage_x[i]\n",
    "#     num = pathway1.split('/')\n",
    "#     percent1 = (int(num[0]) / int(num[1])) *100\n",
    "#     print(percent1)\n",
    "\n",
    "#     pathway2 =  top_10_pathways.Coverage_y[i]\n",
    "#     num = pathway2.split('/')\n",
    "#     percent2 = (int(num[0]) / int(num[1])) *100\n",
    "#     print(percent2)\n",
    "\n",
    "#     percent = (percent1+percent2)/2\n",
    "#     print(percent)\n",
    "#     label.append(\"  \"+ str(round(percent,1))+ '%')\n",
    "\n",
    "# print(label)\n",
    "\n",
    "\n",
    "#Getting pathway coverage by using the Reactome pathway definitions concatenated together\n",
    "for i in range(10):\n",
    "    pathway_nam = top_10_pathways.iloc[i,0]\n",
    "    \n",
    "    num_in_df = len(filtered_dict[pathway_nam])\n",
    "    num_whole_pathway = len(pathways_dict[pathway_nam])\n",
    "\n",
    "    percent = (num_in_df/num_whole_pathway) * 100\n",
    "\n",
    "    label.append(\"  \"+ str(round(percent,1))+ '%')\n",
    "\n",
    "ax.bar_label(ax.containers[0], labels=label)\n",
    "\n",
    "#plt.title('ORA for integrated data',fontsize=22,pad=10)\n",
    "plt.xlabel('Unadjusted p-value',fontsize=26)\n",
    "plt.ylabel('Pathway name',fontsize=26) \n",
    "plt.xlim(0, 0.57);\n",
    "\n",
    "plt.axvline(0.05, c=\"black\")\n",
    "\n",
    "#plt.savefig( '../Figures/integrated_ORA_top_10.png' , dpi=300,bbox_inches = 'tight' , pad_inches = 0.2 , facecolor='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene-set Enrichment Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res = sspa.sspa_gsea(df.iloc[:,:-2], df[\"Group\"], reactome_pathways)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1372</th>\n",
       "      <th>16610</th>\n",
       "      <th>72665</th>\n",
       "      <th>27823</th>\n",
       "      <th>30915</th>\n",
       "      <th>37373</th>\n",
       "      <th>16831</th>\n",
       "      <th>545959</th>\n",
       "      <th>17050</th>\n",
       "      <th>16359</th>\n",
       "      <th>...</th>\n",
       "      <th>133693</th>\n",
       "      <th>133694</th>\n",
       "      <th>28036</th>\n",
       "      <th>28238</th>\n",
       "      <th>76341</th>\n",
       "      <th>89312</th>\n",
       "      <th>17861</th>\n",
       "      <th>89188</th>\n",
       "      <th>WHO_status</th>\n",
       "      <th>Group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>INCOV001</th>\n",
       "      <td>-0.342522</td>\n",
       "      <td>-0.830841</td>\n",
       "      <td>-0.703451</td>\n",
       "      <td>-0.451278</td>\n",
       "      <td>0.153207</td>\n",
       "      <td>-0.443252</td>\n",
       "      <td>-0.228725</td>\n",
       "      <td>0.501998</td>\n",
       "      <td>-0.861886</td>\n",
       "      <td>-0.344155</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.465505</td>\n",
       "      <td>0.055753</td>\n",
       "      <td>-0.225622</td>\n",
       "      <td>0.400398</td>\n",
       "      <td>0.163210</td>\n",
       "      <td>1.579262</td>\n",
       "      <td>-0.162221</td>\n",
       "      <td>-0.603790</td>\n",
       "      <td>3-4</td>\n",
       "      <td>Severe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV002</th>\n",
       "      <td>-0.125585</td>\n",
       "      <td>-1.002246</td>\n",
       "      <td>0.399273</td>\n",
       "      <td>-0.828341</td>\n",
       "      <td>-0.980576</td>\n",
       "      <td>0.373921</td>\n",
       "      <td>-1.076979</td>\n",
       "      <td>-0.131876</td>\n",
       "      <td>-0.869093</td>\n",
       "      <td>-0.467326</td>\n",
       "      <td>...</td>\n",
       "      <td>0.509336</td>\n",
       "      <td>0.627269</td>\n",
       "      <td>0.321718</td>\n",
       "      <td>-0.158514</td>\n",
       "      <td>0.001469</td>\n",
       "      <td>-1.051485</td>\n",
       "      <td>1.299330</td>\n",
       "      <td>0.578560</td>\n",
       "      <td>5-7</td>\n",
       "      <td>Severe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV003</th>\n",
       "      <td>-0.859326</td>\n",
       "      <td>-0.409159</td>\n",
       "      <td>3.573100</td>\n",
       "      <td>-0.546783</td>\n",
       "      <td>-0.993647</td>\n",
       "      <td>1.547649</td>\n",
       "      <td>0.370131</td>\n",
       "      <td>-0.020399</td>\n",
       "      <td>-0.703786</td>\n",
       "      <td>-0.015627</td>\n",
       "      <td>...</td>\n",
       "      <td>2.056610</td>\n",
       "      <td>1.717149</td>\n",
       "      <td>-0.297723</td>\n",
       "      <td>-1.276891</td>\n",
       "      <td>-0.658288</td>\n",
       "      <td>2.141594</td>\n",
       "      <td>-1.091183</td>\n",
       "      <td>0.447197</td>\n",
       "      <td>5-7</td>\n",
       "      <td>Severe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV004</th>\n",
       "      <td>-0.562999</td>\n",
       "      <td>-1.560249</td>\n",
       "      <td>-1.269383</td>\n",
       "      <td>-1.437919</td>\n",
       "      <td>0.108734</td>\n",
       "      <td>-0.192066</td>\n",
       "      <td>0.161591</td>\n",
       "      <td>-0.586759</td>\n",
       "      <td>-1.457246</td>\n",
       "      <td>-0.500612</td>\n",
       "      <td>...</td>\n",
       "      <td>0.793065</td>\n",
       "      <td>-0.033125</td>\n",
       "      <td>-0.366082</td>\n",
       "      <td>-0.642671</td>\n",
       "      <td>1.724436</td>\n",
       "      <td>2.074193</td>\n",
       "      <td>-1.021184</td>\n",
       "      <td>-0.410849</td>\n",
       "      <td>3-4</td>\n",
       "      <td>Severe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV005</th>\n",
       "      <td>-0.741957</td>\n",
       "      <td>0.387909</td>\n",
       "      <td>-0.711036</td>\n",
       "      <td>-0.711515</td>\n",
       "      <td>0.670720</td>\n",
       "      <td>0.610591</td>\n",
       "      <td>-0.322241</td>\n",
       "      <td>-0.793451</td>\n",
       "      <td>-1.230681</td>\n",
       "      <td>-0.246493</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.184967</td>\n",
       "      <td>-0.469553</td>\n",
       "      <td>-0.293869</td>\n",
       "      <td>0.558366</td>\n",
       "      <td>-0.390308</td>\n",
       "      <td>-0.455737</td>\n",
       "      <td>-0.535223</td>\n",
       "      <td>-0.847727</td>\n",
       "      <td>3-4</td>\n",
       "      <td>Severe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV135</th>\n",
       "      <td>-2.075777</td>\n",
       "      <td>-0.485936</td>\n",
       "      <td>-0.671225</td>\n",
       "      <td>0.069483</td>\n",
       "      <td>0.042579</td>\n",
       "      <td>0.357250</td>\n",
       "      <td>-0.505272</td>\n",
       "      <td>-0.378441</td>\n",
       "      <td>-0.444915</td>\n",
       "      <td>-0.368108</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.516557</td>\n",
       "      <td>-0.392716</td>\n",
       "      <td>-0.328170</td>\n",
       "      <td>1.013774</td>\n",
       "      <td>-0.391566</td>\n",
       "      <td>-0.407585</td>\n",
       "      <td>1.041224</td>\n",
       "      <td>-0.866190</td>\n",
       "      <td>1-2</td>\n",
       "      <td>Mild</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV136</th>\n",
       "      <td>-0.174602</td>\n",
       "      <td>1.167644</td>\n",
       "      <td>0.055499</td>\n",
       "      <td>0.113342</td>\n",
       "      <td>-0.663851</td>\n",
       "      <td>0.799367</td>\n",
       "      <td>0.004697</td>\n",
       "      <td>0.748894</td>\n",
       "      <td>1.328753</td>\n",
       "      <td>0.391221</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.741578</td>\n",
       "      <td>-0.730981</td>\n",
       "      <td>-0.380517</td>\n",
       "      <td>-1.251641</td>\n",
       "      <td>0.389163</td>\n",
       "      <td>-0.603048</td>\n",
       "      <td>0.941619</td>\n",
       "      <td>-0.511694</td>\n",
       "      <td>3-4</td>\n",
       "      <td>Severe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV137</th>\n",
       "      <td>0.274900</td>\n",
       "      <td>-0.014841</td>\n",
       "      <td>-0.928512</td>\n",
       "      <td>0.885263</td>\n",
       "      <td>-1.591862</td>\n",
       "      <td>0.230250</td>\n",
       "      <td>0.057818</td>\n",
       "      <td>-0.775962</td>\n",
       "      <td>-0.179700</td>\n",
       "      <td>-0.534786</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.306132</td>\n",
       "      <td>-1.246428</td>\n",
       "      <td>1.763824</td>\n",
       "      <td>1.258387</td>\n",
       "      <td>0.448359</td>\n",
       "      <td>-0.508931</td>\n",
       "      <td>0.268474</td>\n",
       "      <td>-0.997206</td>\n",
       "      <td>5-7</td>\n",
       "      <td>Severe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV139</th>\n",
       "      <td>0.772804</td>\n",
       "      <td>1.408806</td>\n",
       "      <td>1.473315</td>\n",
       "      <td>2.182921</td>\n",
       "      <td>-0.380455</td>\n",
       "      <td>-0.175916</td>\n",
       "      <td>0.376092</td>\n",
       "      <td>0.412578</td>\n",
       "      <td>1.853690</td>\n",
       "      <td>-0.368432</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.050924</td>\n",
       "      <td>-0.909450</td>\n",
       "      <td>0.962331</td>\n",
       "      <td>0.292543</td>\n",
       "      <td>-0.404695</td>\n",
       "      <td>0.919776</td>\n",
       "      <td>-1.226216</td>\n",
       "      <td>0.673867</td>\n",
       "      <td>3-4</td>\n",
       "      <td>Severe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV142</th>\n",
       "      <td>-0.899749</td>\n",
       "      <td>1.791799</td>\n",
       "      <td>-1.411808</td>\n",
       "      <td>-1.034888</td>\n",
       "      <td>-1.556365</td>\n",
       "      <td>0.903904</td>\n",
       "      <td>-0.487092</td>\n",
       "      <td>-0.442150</td>\n",
       "      <td>1.805891</td>\n",
       "      <td>-0.494940</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.246847</td>\n",
       "      <td>-1.201487</td>\n",
       "      <td>-0.416287</td>\n",
       "      <td>0.030042</td>\n",
       "      <td>-0.882021</td>\n",
       "      <td>-0.022782</td>\n",
       "      <td>0.792374</td>\n",
       "      <td>-0.161666</td>\n",
       "      <td>5-7</td>\n",
       "      <td>Severe</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>128 rows × 335 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               1372     16610     72665     27823     30915     37373  \\\n",
       "sample_id                                                               \n",
       "INCOV001  -0.342522 -0.830841 -0.703451 -0.451278  0.153207 -0.443252   \n",
       "INCOV002  -0.125585 -1.002246  0.399273 -0.828341 -0.980576  0.373921   \n",
       "INCOV003  -0.859326 -0.409159  3.573100 -0.546783 -0.993647  1.547649   \n",
       "INCOV004  -0.562999 -1.560249 -1.269383 -1.437919  0.108734 -0.192066   \n",
       "INCOV005  -0.741957  0.387909 -0.711036 -0.711515  0.670720  0.610591   \n",
       "...             ...       ...       ...       ...       ...       ...   \n",
       "INCOV135  -2.075777 -0.485936 -0.671225  0.069483  0.042579  0.357250   \n",
       "INCOV136  -0.174602  1.167644  0.055499  0.113342 -0.663851  0.799367   \n",
       "INCOV137   0.274900 -0.014841 -0.928512  0.885263 -1.591862  0.230250   \n",
       "INCOV139   0.772804  1.408806  1.473315  2.182921 -0.380455 -0.175916   \n",
       "INCOV142  -0.899749  1.791799 -1.411808 -1.034888 -1.556365  0.903904   \n",
       "\n",
       "              16831    545959     17050     16359  ...    133693    133694  \\\n",
       "sample_id                                          ...                       \n",
       "INCOV001  -0.228725  0.501998 -0.861886 -0.344155  ... -0.465505  0.055753   \n",
       "INCOV002  -1.076979 -0.131876 -0.869093 -0.467326  ...  0.509336  0.627269   \n",
       "INCOV003   0.370131 -0.020399 -0.703786 -0.015627  ...  2.056610  1.717149   \n",
       "INCOV004   0.161591 -0.586759 -1.457246 -0.500612  ...  0.793065 -0.033125   \n",
       "INCOV005  -0.322241 -0.793451 -1.230681 -0.246493  ... -0.184967 -0.469553   \n",
       "...             ...       ...       ...       ...  ...       ...       ...   \n",
       "INCOV135  -0.505272 -0.378441 -0.444915 -0.368108  ... -0.516557 -0.392716   \n",
       "INCOV136   0.004697  0.748894  1.328753  0.391221  ... -0.741578 -0.730981   \n",
       "INCOV137   0.057818 -0.775962 -0.179700 -0.534786  ... -1.306132 -1.246428   \n",
       "INCOV139   0.376092  0.412578  1.853690 -0.368432  ... -1.050924 -0.909450   \n",
       "INCOV142  -0.487092 -0.442150  1.805891 -0.494940  ... -1.246847 -1.201487   \n",
       "\n",
       "              28036     28238     76341     89312     17861     89188  \\\n",
       "sample_id                                                               \n",
       "INCOV001  -0.225622  0.400398  0.163210  1.579262 -0.162221 -0.603790   \n",
       "INCOV002   0.321718 -0.158514  0.001469 -1.051485  1.299330  0.578560   \n",
       "INCOV003  -0.297723 -1.276891 -0.658288  2.141594 -1.091183  0.447197   \n",
       "INCOV004  -0.366082 -0.642671  1.724436  2.074193 -1.021184 -0.410849   \n",
       "INCOV005  -0.293869  0.558366 -0.390308 -0.455737 -0.535223 -0.847727   \n",
       "...             ...       ...       ...       ...       ...       ...   \n",
       "INCOV135  -0.328170  1.013774 -0.391566 -0.407585  1.041224 -0.866190   \n",
       "INCOV136  -0.380517 -1.251641  0.389163 -0.603048  0.941619 -0.511694   \n",
       "INCOV137   1.763824  1.258387  0.448359 -0.508931  0.268474 -0.997206   \n",
       "INCOV139   0.962331  0.292543 -0.404695  0.919776 -1.226216  0.673867   \n",
       "INCOV142  -0.416287  0.030042 -0.882021 -0.022782  0.792374 -0.161666   \n",
       "\n",
       "           WHO_status   Group  \n",
       "sample_id                      \n",
       "INCOV001          3-4  Severe  \n",
       "INCOV002          5-7  Severe  \n",
       "INCOV003          5-7  Severe  \n",
       "INCOV004          3-4  Severe  \n",
       "INCOV005          3-4  Severe  \n",
       "...               ...     ...  \n",
       "INCOV135          1-2    Mild  \n",
       "INCOV136          3-4  Severe  \n",
       "INCOV137          5-7  Severe  \n",
       "INCOV139          3-4  Severe  \n",
       "INCOV142          5-7  Severe  \n",
       "\n",
       "[128 rows x 335 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathway_ID</th>\n",
       "      <th>Pathway_name</th>\n",
       "      <th>ES</th>\n",
       "      <th>NES</th>\n",
       "      <th>P-value</th>\n",
       "      <th>P-adjust FDR</th>\n",
       "      <th>P-adjust FWER</th>\n",
       "      <th>Tag %</th>\n",
       "      <th>Entity %</th>\n",
       "      <th>Leading_edge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>R-HSA-425407</td>\n",
       "      <td>SLC-mediated transmembrane transport</td>\n",
       "      <td>0.452285</td>\n",
       "      <td>1.474488</td>\n",
       "      <td>0.051383</td>\n",
       "      <td>0.537223</td>\n",
       "      <td>0.585</td>\n",
       "      <td>15/33</td>\n",
       "      <td>31.23%</td>\n",
       "      <td>4208;16680;16199;16283;15344;16040;15756;17568...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>R-HSA-5663205</td>\n",
       "      <td>Infectious disease</td>\n",
       "      <td>0.911178</td>\n",
       "      <td>1.442434</td>\n",
       "      <td>0.043738</td>\n",
       "      <td>0.555563</td>\n",
       "      <td>0.696</td>\n",
       "      <td>1/2</td>\n",
       "      <td>8.71%</td>\n",
       "      <td>16680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>R-HSA-73614</td>\n",
       "      <td>Pyrimidine salvage</td>\n",
       "      <td>0.757897</td>\n",
       "      <td>1.418139</td>\n",
       "      <td>0.067485</td>\n",
       "      <td>0.566262</td>\n",
       "      <td>0.761</td>\n",
       "      <td>2/4</td>\n",
       "      <td>23.72%</td>\n",
       "      <td>17802;17568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>R-HSA-1614635</td>\n",
       "      <td>Sulfur amino acid metabolism</td>\n",
       "      <td>0.703485</td>\n",
       "      <td>1.451648</td>\n",
       "      <td>0.071571</td>\n",
       "      <td>0.580345</td>\n",
       "      <td>0.666</td>\n",
       "      <td>4/6</td>\n",
       "      <td>26.43%</td>\n",
       "      <td>30831;17509;17482;17750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>R-HSA-2408522</td>\n",
       "      <td>Selenoamino acid metabolism</td>\n",
       "      <td>0.96654</td>\n",
       "      <td>1.48528</td>\n",
       "      <td>0.006237</td>\n",
       "      <td>0.580781</td>\n",
       "      <td>0.55</td>\n",
       "      <td>1/2</td>\n",
       "      <td>3.30%</td>\n",
       "      <td>30831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>R-HSA-2262752</td>\n",
       "      <td>Cellular responses to stress</td>\n",
       "      <td>0.736364</td>\n",
       "      <td>1.16961</td>\n",
       "      <td>0.298419</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>3/3</td>\n",
       "      <td>27.03%</td>\n",
       "      <td>15756;17351;16827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>R-HSA-5619063</td>\n",
       "      <td>Defective SLC29A3 causes histiocytosis-lymphad...</td>\n",
       "      <td>-0.396341</td>\n",
       "      <td>-0.758293</td>\n",
       "      <td>0.737288</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5/5</td>\n",
       "      <td>61.26%</td>\n",
       "      <td>16708;16704;16335;17562;17596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>R-HSA-83936</td>\n",
       "      <td>Transport of nucleosides and free purine and p...</td>\n",
       "      <td>0.539713</td>\n",
       "      <td>1.189375</td>\n",
       "      <td>0.266145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.997</td>\n",
       "      <td>4/8</td>\n",
       "      <td>39.34%</td>\n",
       "      <td>16040;17568;17368;17596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>R-HSA-1483257</td>\n",
       "      <td>Phospholipid metabolism</td>\n",
       "      <td>-0.563098</td>\n",
       "      <td>-1.101136</td>\n",
       "      <td>0.371032</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2/5</td>\n",
       "      <td>3.90%</td>\n",
       "      <td>18132;17553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>R-HSA-71403</td>\n",
       "      <td>Citric acid cycle (TCA cycle)</td>\n",
       "      <td>-0.531722</td>\n",
       "      <td>-0.783552</td>\n",
       "      <td>0.791498</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1/2</td>\n",
       "      <td>29.43%</td>\n",
       "      <td>30769</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>144 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Pathway_ID                                       Pathway_name  \\\n",
       "5     R-HSA-425407               SLC-mediated transmembrane transport   \n",
       "8    R-HSA-5663205                                 Infectious disease   \n",
       "10     R-HSA-73614                                 Pyrimidine salvage   \n",
       "6    R-HSA-1614635                       Sulfur amino acid metabolism   \n",
       "3    R-HSA-2408522                        Selenoamino acid metabolism   \n",
       "..             ...                                                ...   \n",
       "32   R-HSA-2262752                       Cellular responses to stress   \n",
       "115  R-HSA-5619063  Defective SLC29A3 causes histiocytosis-lymphad...   \n",
       "31     R-HSA-83936  Transport of nucleosides and free purine and p...   \n",
       "39   R-HSA-1483257                            Phospholipid metabolism   \n",
       "111    R-HSA-71403                      Citric acid cycle (TCA cycle)   \n",
       "\n",
       "           ES       NES   P-value P-adjust FDR P-adjust FWER  Tag % Entity %  \\\n",
       "5    0.452285  1.474488  0.051383     0.537223         0.585  15/33   31.23%   \n",
       "8    0.911178  1.442434  0.043738     0.555563         0.696    1/2    8.71%   \n",
       "10   0.757897  1.418139  0.067485     0.566262         0.761    2/4   23.72%   \n",
       "6    0.703485  1.451648  0.071571     0.580345         0.666    4/6   26.43%   \n",
       "3     0.96654   1.48528  0.006237     0.580781          0.55    1/2    3.30%   \n",
       "..        ...       ...       ...          ...           ...    ...      ...   \n",
       "32   0.736364   1.16961  0.298419          1.0         0.999    3/3   27.03%   \n",
       "115 -0.396341 -0.758293  0.737288          1.0           1.0    5/5   61.26%   \n",
       "31   0.539713  1.189375  0.266145          1.0         0.997    4/8   39.34%   \n",
       "39  -0.563098 -1.101136  0.371032          1.0           1.0    2/5    3.90%   \n",
       "111 -0.531722 -0.783552  0.791498          1.0           1.0    1/2   29.43%   \n",
       "\n",
       "                                          Leading_edge  \n",
       "5    4208;16680;16199;16283;15344;16040;15756;17568...  \n",
       "8                                                16680  \n",
       "10                                         17802;17568  \n",
       "6                              30831;17509;17482;17750  \n",
       "3                                                30831  \n",
       "..                                                 ...  \n",
       "32                                   15756;17351;16827  \n",
       "115                      16708;16704;16335;17562;17596  \n",
       "31                             16040;17568;17368;17596  \n",
       "39                                         18132;17553  \n",
       "111                                              30769  \n",
       "\n",
       "[144 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(gsea_res.sort_values(by=\"P-adjust FDR\")) #0.54 is the lowest for metabolomic, 0.056 for proteomic\n",
    "#display(gsea_res.sort_values(by=\"P-value\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Change name to have a line break otherwise it won't fit on the plot\n",
    "gsea_res.Pathway_name[0] = 'Regulation of Insulin-like Growth Factor (IGF) transport and \\n uptake by Insulin-like Growth Factor Binding Proteins (IGFBPs)'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_10_pathways_gsea"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.lines import Line2D #To colour the bar by enrichment\n",
    "\n",
    "top_10_pathways_gsea = gsea_res.sort_values(by=\"P-value\").iloc[0:10, :]\n",
    "plt.figure(figsize=(10, 7))\n",
    "\n",
    "sns.set(font_scale = 2) #sns.set(font_scale = 1.2) for metabolomic because of long label\n",
    "sns.set_style(\"ticks\") # same as \"white\" but with ticks\n",
    "\n",
    "# set bar colour based on normalised enrichment score sign\n",
    "bar_color = ['tab:red' if float(i) > 0 else 'tab:blue' for i in top_10_pathways_gsea['NES']]\n",
    "ax = sns.barplot(data=top_10_pathways_gsea, y=\"Pathway_name\", x=\"P-value\", orient=\"h\", palette=bar_color)\n",
    "\n",
    "#Increase space between the bars and the axis\n",
    "ax.relim()\n",
    "ax.autoscale_view()\n",
    "\n",
    "\n",
    "\n",
    "label = []\n",
    "for pathway in top_10_pathways_gsea[\"Entity %\"]:\n",
    "    label.append(\"  \"+ pathway)\n",
    "\n",
    "print(label)\n",
    "\n",
    "ax.bar_label(ax.containers[0], labels=label)\n",
    "\n",
    "#plt.title('GSEA for proteomic data',fontsize=22,pad=10)\n",
    "plt.xlabel('Unadjusted p-value',fontsize=26)\n",
    "plt.ylabel('Pathway name',fontsize=26) ;\n",
    "\n",
    "plt.axvline(0.05, c=\"black\")\n",
    "\n",
    "# add legend\n",
    "custom_lines = [Line2D([0], [0], color='tab:red', lw=4),\n",
    "                Line2D([0], [0], color='tab:blue', lw=4)]\n",
    "plt.legend(handles=custom_lines, labels=['Positive enrichment score', 'Negative enrichment score'],loc=\"upper right\")\n",
    "\n",
    "#plt.savefig( '../Figures/metabolomic_GSEA_top_10.png' , dpi=300,bbox_inches = 'tight' , pad_inches = 0.2 , facecolor='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For integrated data, take the two GSEA results and combine the p-values with Fisher's method, which is what is commonly done (See Maghsoudi et al., 2021 for examples):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabolomic_gsea = gsea_res.sort_values(by=\"P-value\")\n",
    "metabolomic_gsea = metabolomic_gsea.set_index(['Pathway_ID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteomic_gsea = gsea_res.sort_values(by=\"P-value\")\n",
    "proteomic_gsea = proteomic_gsea.set_index(['Pathway_ID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabolomic_gsea"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteomic_gsea"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result = metabolomic_gsea.merge(proteomic_gsea, how='inner',right_index=True, left_index = True)\n",
    "result[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result[\"combined_pval\"] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(len(result.index)):\n",
    "    ID_list = [result[\"P-value_x\"][i],result[\"P-value_y\"][i]]\n",
    "    print(ID_list)\n",
    "    test_stat,pval = scipy.stats.combine_pvalues(ID_list, method='fisher',weights=None)\n",
    "    print(pval)\n",
    "    result[\"combined_pval\"][i] = pval\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Obtain pathway coverage for the top 10 results\n",
    "df = pd.read_csv(\"../Data/Su_integrated_data.csv\", index_col=0)\n",
    "reactome_pathways = pd.read_csv(\"../Data/Reactome_multi_omics_ChEBI_Uniprot.csv\", index_col=0,dtype=\"str\") #Dtype warning because in some columns, some values are in string format whereas some are in integer format, that's why I specify dtype=\"str\"\n",
    "kpca_scores = sspa.sspa_kpca(df.iloc[:,:-2], reactome_pathways)\n",
    "\n",
    "\n",
    "#Convert pathway ID to name\n",
    "root_path = pd.read_excel('../Data/Root_pathways.xlsx', header=None)\n",
    "root_pathway_dict = {root_path[0][i]:root_path[1][i] for i in range(0,len(root_path))}\n",
    "\n",
    "root_pathway_names = list(root_pathway_dict.keys())\n",
    "#Using Sara's code, remove root pathways\n",
    "kpca_scores = kpca_scores.drop(columns = list(set(root_pathway_names) & set(kpca_scores.columns)))\n",
    "\n",
    "\n",
    "\n",
    "#Filter out the molecules in the pathways that are not present in the dataset\n",
    "#Obtain all unique values in dataset\n",
    "compounds_present = list(df.columns[:-2])\n",
    "filtered_dict = {} \n",
    "\n",
    "#Obtain pathways and corresponding molecules for all Reactome pathways, store as dictionary\n",
    "orig_dict = sspa.utils.pathwaydf_to_dict(reactome_pathways)\n",
    "\n",
    "#Filter out dictionary to retain only the pathways that remain after kPCA\n",
    "my_keys = kpca_scores.columns\n",
    "pathways_dict = {key: orig_dict[key] for key in my_keys}\n",
    "\n",
    "\n",
    "#My code adapted from Cecilia's\n",
    "#If the key values are not part of the compounds in dataset then remove\n",
    "for key,value in pathways_dict.items():\n",
    "    new_val = [item for item in value if item in compounds_present]\n",
    "    if len(new_val) >= 2: #at least two compounds in the pathway\n",
    "        filtered_dict[key] = new_val"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.lines import Line2D\n",
    "\n",
    "top_10_pathways_gsea = result.sort_values(by=\"combined_pval\").iloc[0:10, :]\n",
    "plt.figure(figsize=(10, 7))\n",
    "\n",
    "sns.set(font_scale = 2) #sns.set(font_scale = 1.2) for metabolomic because of long label\n",
    "sns.set_style(\"ticks\") # same as \"white\" but with ticks\n",
    "\n",
    "# set bar colour based on normalised enrichment score sign\n",
    "bar_color = ['tab:red' if float(i) > 0 else 'tab:blue' for i in top_10_pathways_gsea['NES_x']]\n",
    "ax = sns.barplot(data=top_10_pathways_gsea, y=\"Pathway_name_x\", x=\"combined_pval\", orient=\"h\", palette=bar_color)\n",
    "\n",
    "#Increase space between the bars and the axis\n",
    "ax.relim()\n",
    "ax.autoscale_view()\n",
    "\n",
    "\n",
    "label=[]\n",
    "# for i in range(10):\n",
    "#     pathway1 =  top_10_pathways_gsea[\"Entity %_x\"][i][:-1]\n",
    "    \n",
    "#     pathway2 =  top_10_pathways_gsea[\"Entity %_y\"][i][:-1]\n",
    "\n",
    "#     percent = (float(pathway1)+float(pathway2))/2\n",
    "#     print(percent)\n",
    "#     label.append(\"  \"+ str(round(percent,1))+ '%')\n",
    "\n",
    "# print(label)\n",
    "\n",
    "for i in range(10):\n",
    "\n",
    "    pathway_nam = top_10_pathways_gsea.index[i]\n",
    "    \n",
    "    num_in_df = len(filtered_dict[pathway_nam])\n",
    "    num_whole_pathway = len(pathways_dict[pathway_nam])\n",
    "\n",
    "    percent = (num_in_df/num_whole_pathway) * 100\n",
    "\n",
    "    label.append(\"  \"+ str(round(percent,1))+ '%')\n",
    "\n",
    "\n",
    "print(label)\n",
    "\n",
    "\n",
    "\n",
    "ax.bar_label(ax.containers[0], labels=label)\n",
    "\n",
    "#plt.title('GSEA for integrated data',fontsize=22, pad=10)\n",
    "plt.xlabel('Unadjusted p-value',fontsize=26)\n",
    "plt.ylabel('Pathway name',fontsize=26) \n",
    "plt.xlim(0, 0.21);\n",
    "\n",
    "plt.axvline(0.05, c=\"black\")\n",
    "\n",
    "#Add legend\n",
    "custom_lines = [Line2D([0], [0], color='tab:red', lw=4),\n",
    "                Line2D([0], [0], color='tab:blue', lw=4)]\n",
    "plt.legend(handles=custom_lines, labels=['Positive enrichment score', 'Negative enrichment score'],loc=\"upper right\")\n",
    "\n",
    "\n",
    "#plt.savefig( '../Figures/integrated_GSEA_top_10.png' , dpi=300,bbox_inches = 'tight' , pad_inches = 0.2 , facecolor='w')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Imperial_Project2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
