{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#miss forest doesn't work for current version of sklearn\n",
    "#if necessary, downgrade sklearn version\n",
    "#https://github.com/scikit-learn/scikit-learn/discussions/25745\n",
    "\n",
    "#pip uninstall scikit-learn -y\n",
    "#pip install scikit-learn==1.1.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.2.4\n"
     ]
    }
   ],
   "source": [
    "import sspa\n",
    "import scipy\n",
    "print(sspa.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "#ignores future warnings for miss forest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import openpyxl   #to read in a sheet of the excel file\n",
    "\n",
    "#imputation libraries  (import sklearn before importing sys and missingpy otherwise an error will happen)\n",
    "import sklearn\n",
    "\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#imputation libraries (read this, then the cell below, then this cell again)\n",
    "import sys\n",
    "sys.modules['sklearn.neighbors.base'] = sklearn.neighbors._base"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'missingpy'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\jenna\\Documents\\Imperial\\Project_2\\Code\\other_scripts\\protein_pre_processing.ipynb Cell 6\u001b[0m line \u001b[0;36m2\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/jenna/Documents/Imperial/Project_2/Code/other_scripts/protein_pre_processing.ipynb#W5sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39m#imputation libraries\u001b[39;00m\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/jenna/Documents/Imperial/Project_2/Code/other_scripts/protein_pre_processing.ipynb#W5sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mmissingpy\u001b[39;00m \u001b[39mimport\u001b[39;00m MissForest\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'missingpy'"
     ]
    }
   ],
   "source": [
    "#imputation libraries\n",
    "from missingpy import MissForest"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Manipulate dataframe"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data can be downloaded for the plasma proteomic dataset from the following paper: <br>\n",
    "Multi-Omics Resolves a Sharp Disease-State Shift between Mild and Moderate COVID-19 (Su et al., 2020)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\jenna\\AppData\\Local\\Temp\\ipykernel_10664\\1282734836.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  first_draws.loc[:,'sample_id'] = first_draws.loc[:,'sample_id'].str.rstrip('-BL')  #remove 'BL' label from the cases (so I can match to WHO status)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lectin, galactoside-binding, soluble, 9</th>\n",
       "      <th>tumor necrosis factor receptor superfamily, member 10a</th>\n",
       "      <th>agouti related protein homolog (mouse)</th>\n",
       "      <th>tumor necrosis factor receptor superfamily, member 10b</th>\n",
       "      <th>tumor necrosis factor receptor superfamily, member 13B</th>\n",
       "      <th>c-fos induced growth factor (vascular endothelial growth factor D)</th>\n",
       "      <th>dickkopf WNT signaling pathway inhibitor 1</th>\n",
       "      <th>renin</th>\n",
       "      <th>platelet-derived growth factor beta polypeptide</th>\n",
       "      <th>growth hormone 1</th>\n",
       "      <th>...</th>\n",
       "      <th>EGF-like-domain, multiple 7</th>\n",
       "      <th>REST corepressor 1</th>\n",
       "      <th>carbonic anhydrase XIV</th>\n",
       "      <th>syntaxin 8</th>\n",
       "      <th>protein kinase, AMP-activated, beta 1 non-catalytic subunit</th>\n",
       "      <th>mitogen-activated protein kinase kinase kinase kinase 5</th>\n",
       "      <th>negative regulator of ubiquitin-like proteins 1</th>\n",
       "      <th>ectonucleoside triphosphate diphosphohydrolase 2</th>\n",
       "      <th>melanoma antigen family D, 1</th>\n",
       "      <th>G protein-coupled receptor 56</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9.467005</td>\n",
       "      <td>3.150612</td>\n",
       "      <td>4.765860</td>\n",
       "      <td>5.853070</td>\n",
       "      <td>9.776093</td>\n",
       "      <td>8.076783</td>\n",
       "      <td>10.286002</td>\n",
       "      <td>7.445927</td>\n",
       "      <td>11.845645</td>\n",
       "      <td>10.276995</td>\n",
       "      <td>...</td>\n",
       "      <td>5.089420</td>\n",
       "      <td>1.904315</td>\n",
       "      <td>2.354585</td>\n",
       "      <td>3.370510</td>\n",
       "      <td>1.457120</td>\n",
       "      <td>6.393335</td>\n",
       "      <td>1.888595</td>\n",
       "      <td>1.272365</td>\n",
       "      <td>-1.015575</td>\n",
       "      <td>1.774740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.414755</td>\n",
       "      <td>1.457353</td>\n",
       "      <td>1.244160</td>\n",
       "      <td>0.712265</td>\n",
       "      <td>2.301213</td>\n",
       "      <td>0.910362</td>\n",
       "      <td>-0.073302</td>\n",
       "      <td>1.053592</td>\n",
       "      <td>0.319995</td>\n",
       "      <td>0.783655</td>\n",
       "      <td>...</td>\n",
       "      <td>5.737120</td>\n",
       "      <td>1.706265</td>\n",
       "      <td>2.174810</td>\n",
       "      <td>2.320480</td>\n",
       "      <td>0.754500</td>\n",
       "      <td>3.058775</td>\n",
       "      <td>0.745930</td>\n",
       "      <td>1.413305</td>\n",
       "      <td>-1.683045</td>\n",
       "      <td>1.542735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9.581175</td>\n",
       "      <td>3.121713</td>\n",
       "      <td>4.649250</td>\n",
       "      <td>5.796290</td>\n",
       "      <td>9.494383</td>\n",
       "      <td>8.618412</td>\n",
       "      <td>9.356512</td>\n",
       "      <td>7.023788</td>\n",
       "      <td>10.991955</td>\n",
       "      <td>9.868815</td>\n",
       "      <td>...</td>\n",
       "      <td>5.046855</td>\n",
       "      <td>1.923825</td>\n",
       "      <td>2.362200</td>\n",
       "      <td>3.534835</td>\n",
       "      <td>1.649450</td>\n",
       "      <td>5.595935</td>\n",
       "      <td>2.099020</td>\n",
       "      <td>1.539440</td>\n",
       "      <td>-0.599360</td>\n",
       "      <td>1.543860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9.391395</td>\n",
       "      <td>3.068790</td>\n",
       "      <td>4.669920</td>\n",
       "      <td>5.773255</td>\n",
       "      <td>11.887820</td>\n",
       "      <td>7.702655</td>\n",
       "      <td>9.278815</td>\n",
       "      <td>8.153380</td>\n",
       "      <td>10.270445</td>\n",
       "      <td>5.840475</td>\n",
       "      <td>...</td>\n",
       "      <td>4.665445</td>\n",
       "      <td>1.910215</td>\n",
       "      <td>2.364895</td>\n",
       "      <td>3.594625</td>\n",
       "      <td>1.468635</td>\n",
       "      <td>5.781550</td>\n",
       "      <td>1.864265</td>\n",
       "      <td>1.657420</td>\n",
       "      <td>-0.094040</td>\n",
       "      <td>1.229660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9.142025</td>\n",
       "      <td>3.198663</td>\n",
       "      <td>5.430080</td>\n",
       "      <td>5.614800</td>\n",
       "      <td>9.746692</td>\n",
       "      <td>8.250982</td>\n",
       "      <td>9.407002</td>\n",
       "      <td>7.645927</td>\n",
       "      <td>11.169755</td>\n",
       "      <td>8.247015</td>\n",
       "      <td>...</td>\n",
       "      <td>5.510625</td>\n",
       "      <td>1.770395</td>\n",
       "      <td>2.533960</td>\n",
       "      <td>3.393515</td>\n",
       "      <td>1.293010</td>\n",
       "      <td>5.925985</td>\n",
       "      <td>1.124380</td>\n",
       "      <td>1.383710</td>\n",
       "      <td>-0.680310</td>\n",
       "      <td>1.763090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>9.786717</td>\n",
       "      <td>3.699688</td>\n",
       "      <td>4.876522</td>\n",
       "      <td>6.064768</td>\n",
       "      <td>9.259790</td>\n",
       "      <td>7.528678</td>\n",
       "      <td>11.526053</td>\n",
       "      <td>7.279940</td>\n",
       "      <td>12.224605</td>\n",
       "      <td>8.855252</td>\n",
       "      <td>...</td>\n",
       "      <td>6.731739</td>\n",
       "      <td>5.465761</td>\n",
       "      <td>2.660046</td>\n",
       "      <td>6.218315</td>\n",
       "      <td>4.499703</td>\n",
       "      <td>9.236677</td>\n",
       "      <td>5.852625</td>\n",
       "      <td>1.008050</td>\n",
       "      <td>1.686931</td>\n",
       "      <td>1.697320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>10.382318</td>\n",
       "      <td>3.511383</td>\n",
       "      <td>4.385330</td>\n",
       "      <td>5.207974</td>\n",
       "      <td>9.330641</td>\n",
       "      <td>7.803160</td>\n",
       "      <td>10.952554</td>\n",
       "      <td>8.125226</td>\n",
       "      <td>12.180945</td>\n",
       "      <td>7.589490</td>\n",
       "      <td>...</td>\n",
       "      <td>6.131039</td>\n",
       "      <td>3.616375</td>\n",
       "      <td>3.294628</td>\n",
       "      <td>5.700506</td>\n",
       "      <td>3.362275</td>\n",
       "      <td>8.424453</td>\n",
       "      <td>4.588876</td>\n",
       "      <td>1.053973</td>\n",
       "      <td>0.996480</td>\n",
       "      <td>1.344471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>10.908705</td>\n",
       "      <td>4.361646</td>\n",
       "      <td>5.858950</td>\n",
       "      <td>6.801220</td>\n",
       "      <td>10.706707</td>\n",
       "      <td>6.956506</td>\n",
       "      <td>11.226867</td>\n",
       "      <td>9.191583</td>\n",
       "      <td>11.765794</td>\n",
       "      <td>4.034886</td>\n",
       "      <td>...</td>\n",
       "      <td>6.102758</td>\n",
       "      <td>3.627333</td>\n",
       "      <td>1.468618</td>\n",
       "      <td>4.953478</td>\n",
       "      <td>3.224846</td>\n",
       "      <td>8.259428</td>\n",
       "      <td>3.961976</td>\n",
       "      <td>0.972515</td>\n",
       "      <td>0.772288</td>\n",
       "      <td>3.442974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>9.624015</td>\n",
       "      <td>3.330472</td>\n",
       "      <td>4.729950</td>\n",
       "      <td>4.996068</td>\n",
       "      <td>9.785855</td>\n",
       "      <td>7.849917</td>\n",
       "      <td>10.436921</td>\n",
       "      <td>7.818691</td>\n",
       "      <td>11.536490</td>\n",
       "      <td>7.558431</td>\n",
       "      <td>...</td>\n",
       "      <td>5.427625</td>\n",
       "      <td>2.387093</td>\n",
       "      <td>1.869467</td>\n",
       "      <td>5.393703</td>\n",
       "      <td>3.315091</td>\n",
       "      <td>8.689472</td>\n",
       "      <td>3.920355</td>\n",
       "      <td>1.406586</td>\n",
       "      <td>0.873741</td>\n",
       "      <td>1.418509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>11.174345</td>\n",
       "      <td>4.703725</td>\n",
       "      <td>7.628121</td>\n",
       "      <td>9.758589</td>\n",
       "      <td>10.389508</td>\n",
       "      <td>7.250206</td>\n",
       "      <td>11.577085</td>\n",
       "      <td>9.109854</td>\n",
       "      <td>12.079503</td>\n",
       "      <td>5.333670</td>\n",
       "      <td>...</td>\n",
       "      <td>7.083230</td>\n",
       "      <td>3.670221</td>\n",
       "      <td>2.719823</td>\n",
       "      <td>6.030698</td>\n",
       "      <td>3.917369</td>\n",
       "      <td>9.252264</td>\n",
       "      <td>4.815057</td>\n",
       "      <td>1.007878</td>\n",
       "      <td>1.571172</td>\n",
       "      <td>3.175467</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>259 rows × 464 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     lectin, galactoside-binding, soluble, 9  \\\n",
       "0                                   9.467005   \n",
       "1                                   2.414755   \n",
       "2                                   9.581175   \n",
       "3                                   9.391395   \n",
       "4                                   9.142025   \n",
       "..                                       ...   \n",
       "254                                 9.786717   \n",
       "255                                10.382318   \n",
       "256                                10.908705   \n",
       "257                                 9.624015   \n",
       "258                                11.174345   \n",
       "\n",
       "     tumor necrosis factor receptor superfamily, member 10a  \\\n",
       "0                                             3.150612        \n",
       "1                                             1.457353        \n",
       "2                                             3.121713        \n",
       "3                                             3.068790        \n",
       "4                                             3.198663        \n",
       "..                                                 ...        \n",
       "254                                           3.699688        \n",
       "255                                           3.511383        \n",
       "256                                           4.361646        \n",
       "257                                           3.330472        \n",
       "258                                           4.703725        \n",
       "\n",
       "     agouti related protein homolog (mouse)  \\\n",
       "0                                  4.765860   \n",
       "1                                  1.244160   \n",
       "2                                  4.649250   \n",
       "3                                  4.669920   \n",
       "4                                  5.430080   \n",
       "..                                      ...   \n",
       "254                                4.876522   \n",
       "255                                4.385330   \n",
       "256                                5.858950   \n",
       "257                                4.729950   \n",
       "258                                7.628121   \n",
       "\n",
       "     tumor necrosis factor receptor superfamily, member 10b  \\\n",
       "0                                             5.853070        \n",
       "1                                             0.712265        \n",
       "2                                             5.796290        \n",
       "3                                             5.773255        \n",
       "4                                             5.614800        \n",
       "..                                                 ...        \n",
       "254                                           6.064768        \n",
       "255                                           5.207974        \n",
       "256                                           6.801220        \n",
       "257                                           4.996068        \n",
       "258                                           9.758589        \n",
       "\n",
       "     tumor necrosis factor receptor superfamily, member 13B  \\\n",
       "0                                             9.776093        \n",
       "1                                             2.301213        \n",
       "2                                             9.494383        \n",
       "3                                            11.887820        \n",
       "4                                             9.746692        \n",
       "..                                                 ...        \n",
       "254                                           9.259790        \n",
       "255                                           9.330641        \n",
       "256                                          10.706707        \n",
       "257                                           9.785855        \n",
       "258                                          10.389508        \n",
       "\n",
       "     c-fos induced growth factor (vascular endothelial growth factor D)  \\\n",
       "0                                             8.076783                    \n",
       "1                                             0.910362                    \n",
       "2                                             8.618412                    \n",
       "3                                             7.702655                    \n",
       "4                                             8.250982                    \n",
       "..                                                 ...                    \n",
       "254                                           7.528678                    \n",
       "255                                           7.803160                    \n",
       "256                                           6.956506                    \n",
       "257                                           7.849917                    \n",
       "258                                           7.250206                    \n",
       "\n",
       "     dickkopf WNT signaling pathway inhibitor 1     renin  \\\n",
       "0                                     10.286002  7.445927   \n",
       "1                                     -0.073302  1.053592   \n",
       "2                                      9.356512  7.023788   \n",
       "3                                      9.278815  8.153380   \n",
       "4                                      9.407002  7.645927   \n",
       "..                                          ...       ...   \n",
       "254                                   11.526053  7.279940   \n",
       "255                                   10.952554  8.125226   \n",
       "256                                   11.226867  9.191583   \n",
       "257                                   10.436921  7.818691   \n",
       "258                                   11.577085  9.109854   \n",
       "\n",
       "     platelet-derived growth factor beta polypeptide  growth hormone 1  ...  \\\n",
       "0                                          11.845645         10.276995  ...   \n",
       "1                                           0.319995          0.783655  ...   \n",
       "2                                          10.991955          9.868815  ...   \n",
       "3                                          10.270445          5.840475  ...   \n",
       "4                                          11.169755          8.247015  ...   \n",
       "..                                               ...               ...  ...   \n",
       "254                                        12.224605          8.855252  ...   \n",
       "255                                        12.180945          7.589490  ...   \n",
       "256                                        11.765794          4.034886  ...   \n",
       "257                                        11.536490          7.558431  ...   \n",
       "258                                        12.079503          5.333670  ...   \n",
       "\n",
       "     EGF-like-domain, multiple 7  REST corepressor 1  carbonic anhydrase XIV  \\\n",
       "0                       5.089420            1.904315                2.354585   \n",
       "1                       5.737120            1.706265                2.174810   \n",
       "2                       5.046855            1.923825                2.362200   \n",
       "3                       4.665445            1.910215                2.364895   \n",
       "4                       5.510625            1.770395                2.533960   \n",
       "..                           ...                 ...                     ...   \n",
       "254                     6.731739            5.465761                2.660046   \n",
       "255                     6.131039            3.616375                3.294628   \n",
       "256                     6.102758            3.627333                1.468618   \n",
       "257                     5.427625            2.387093                1.869467   \n",
       "258                     7.083230            3.670221                2.719823   \n",
       "\n",
       "     syntaxin 8  protein kinase, AMP-activated, beta 1 non-catalytic subunit  \\\n",
       "0      3.370510                                           1.457120             \n",
       "1      2.320480                                           0.754500             \n",
       "2      3.534835                                           1.649450             \n",
       "3      3.594625                                           1.468635             \n",
       "4      3.393515                                           1.293010             \n",
       "..          ...                                                ...             \n",
       "254    6.218315                                           4.499703             \n",
       "255    5.700506                                           3.362275             \n",
       "256    4.953478                                           3.224846             \n",
       "257    5.393703                                           3.315091             \n",
       "258    6.030698                                           3.917369             \n",
       "\n",
       "     mitogen-activated protein kinase kinase kinase kinase 5  \\\n",
       "0                                             6.393335         \n",
       "1                                             3.058775         \n",
       "2                                             5.595935         \n",
       "3                                             5.781550         \n",
       "4                                             5.925985         \n",
       "..                                                 ...         \n",
       "254                                           9.236677         \n",
       "255                                           8.424453         \n",
       "256                                           8.259428         \n",
       "257                                           8.689472         \n",
       "258                                           9.252264         \n",
       "\n",
       "     negative regulator of ubiquitin-like proteins 1  \\\n",
       "0                                           1.888595   \n",
       "1                                           0.745930   \n",
       "2                                           2.099020   \n",
       "3                                           1.864265   \n",
       "4                                           1.124380   \n",
       "..                                               ...   \n",
       "254                                         5.852625   \n",
       "255                                         4.588876   \n",
       "256                                         3.961976   \n",
       "257                                         3.920355   \n",
       "258                                         4.815057   \n",
       "\n",
       "     ectonucleoside triphosphate diphosphohydrolase 2  \\\n",
       "0                                            1.272365   \n",
       "1                                            1.413305   \n",
       "2                                            1.539440   \n",
       "3                                            1.657420   \n",
       "4                                            1.383710   \n",
       "..                                                ...   \n",
       "254                                          1.008050   \n",
       "255                                          1.053973   \n",
       "256                                          0.972515   \n",
       "257                                          1.406586   \n",
       "258                                          1.007878   \n",
       "\n",
       "     melanoma antigen family D, 1  G protein-coupled receptor 56  \n",
       "0                       -1.015575                       1.774740  \n",
       "1                       -1.683045                       1.542735  \n",
       "2                       -0.599360                       1.543860  \n",
       "3                       -0.094040                       1.229660  \n",
       "4                       -0.680310                       1.763090  \n",
       "..                            ...                            ...  \n",
       "254                      1.686931                       1.697320  \n",
       "255                      0.996480                       1.344471  \n",
       "256                      0.772288                       3.442974  \n",
       "257                      0.873741                       1.418509  \n",
       "258                      1.571172                       3.175467  \n",
       "\n",
       "[259 rows x 464 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data = pd.read_excel('../Data/Su_raw.xlsx', sheet_name=3)   #Sheet name = 3, proteomic     (sheet name = 4, metabolomic)\n",
    "\n",
    "controls = data.loc[data['Group'] == 'Healthy Donor ']\n",
    "cases = data.loc[data['Group'] == 'COVID19 ']\n",
    "\n",
    "#There are two readings for each patient e.g. INCOV051-BL and INCOV051-AC\n",
    "#BL stands for the first draw (this is the one we want); AC is the second draw and can be filtered out\n",
    "first_draws = cases[cases.loc[:,'sample_id'].str.contains(\"BL\")]\n",
    "first_draws.loc[:,'sample_id'] = first_draws.loc[:,'sample_id'].str.rstrip('-BL')  #remove 'BL' label from the cases (so I can match to WHO status)\n",
    "\n",
    "df = pd.concat([controls, first_draws], ignore_index=True)  #concatenate dataset and reset index \n",
    "df.insert(0, \"WHO_status\", '0') #insert new column for WHO_status and temporarily add all zeros\n",
    "\n",
    "#display(df['sample_id'][124:130])    #first instance of a case is at index 124\n",
    "display(df.iloc[:,3:])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study Subject ID</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Blood draw time point</th>\n",
       "      <th>Patient Location</th>\n",
       "      <th>Who Ordinal Scale</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Race</th>\n",
       "      <th>BMI</th>\n",
       "      <th>...</th>\n",
       "      <th>Respiratory Support</th>\n",
       "      <th>Temperature</th>\n",
       "      <th>Systolic BP</th>\n",
       "      <th>Diastolic BP</th>\n",
       "      <th>10X?</th>\n",
       "      <th>Metabolomics</th>\n",
       "      <th>Proteomics?</th>\n",
       "      <th>Isoplexis CD8 assay?</th>\n",
       "      <th>Isoplexis CD4 assay?</th>\n",
       "      <th>Isoplexis Monocyte assay?</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>INCOV001</td>\n",
       "      <td>INCOV001-1</td>\n",
       "      <td>T1</td>\n",
       "      <td>Hospital</td>\n",
       "      <td>3</td>\n",
       "      <td>Female</td>\n",
       "      <td>77</td>\n",
       "      <td>NOT Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>33.657783</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>36.6</td>\n",
       "      <td>139.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>INCOV002</td>\n",
       "      <td>INCOV002-1</td>\n",
       "      <td>T1</td>\n",
       "      <td>ICU</td>\n",
       "      <td>5</td>\n",
       "      <td>Male</td>\n",
       "      <td>39</td>\n",
       "      <td>NOT Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>27.900747</td>\n",
       "      <td>...</td>\n",
       "      <td>High flow nasal cannula (HFNC)</td>\n",
       "      <td>37.3</td>\n",
       "      <td>96.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>INCOV003</td>\n",
       "      <td>INCOV003-1</td>\n",
       "      <td>T1</td>\n",
       "      <td>ICU</td>\n",
       "      <td>7</td>\n",
       "      <td>Male</td>\n",
       "      <td>64</td>\n",
       "      <td>NOT Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>24.546988</td>\n",
       "      <td>...</td>\n",
       "      <td>Other</td>\n",
       "      <td>37.2</td>\n",
       "      <td>89.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>INCOV004</td>\n",
       "      <td>INCOV004-1</td>\n",
       "      <td>T1</td>\n",
       "      <td>Hospital</td>\n",
       "      <td>4</td>\n",
       "      <td>Male</td>\n",
       "      <td>76</td>\n",
       "      <td>NOT Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>23.477620</td>\n",
       "      <td>...</td>\n",
       "      <td>Nasal cannula</td>\n",
       "      <td>36.2</td>\n",
       "      <td>137.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>INCOV005</td>\n",
       "      <td>INCOV005-1</td>\n",
       "      <td>T1</td>\n",
       "      <td>Hospital</td>\n",
       "      <td>4</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>NOT Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>29.703121</td>\n",
       "      <td>...</td>\n",
       "      <td>Nasal cannula</td>\n",
       "      <td>36.8</td>\n",
       "      <td>126.0</td>\n",
       "      <td>105.0</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>INCOV141</td>\n",
       "      <td>INCOV141-1</td>\n",
       "      <td>T1</td>\n",
       "      <td>Home (mobile phlebotomy)</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>31</td>\n",
       "      <td>NOT Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>26.575066</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>INCOV142</td>\n",
       "      <td>INCOV142-1</td>\n",
       "      <td>T1</td>\n",
       "      <td>ICU</td>\n",
       "      <td>5</td>\n",
       "      <td>Male</td>\n",
       "      <td>64</td>\n",
       "      <td>NOT Hispanic or Latino</td>\n",
       "      <td>Asian</td>\n",
       "      <td>30.980420</td>\n",
       "      <td>...</td>\n",
       "      <td>High flow nasal cannula (HFNC)</td>\n",
       "      <td>36.2</td>\n",
       "      <td>104.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>INCOV144</td>\n",
       "      <td>INCOV144-1</td>\n",
       "      <td>T1</td>\n",
       "      <td>Home (mobile phlebotomy)</td>\n",
       "      <td>1 or 2</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>NOT Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>INCOV145</td>\n",
       "      <td>INCOV145-1</td>\n",
       "      <td>T1</td>\n",
       "      <td>Hospital</td>\n",
       "      <td>4</td>\n",
       "      <td>Male</td>\n",
       "      <td>55</td>\n",
       "      <td>NOT Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>35.871500</td>\n",
       "      <td>...</td>\n",
       "      <td>Nasal cannula</td>\n",
       "      <td>36.9</td>\n",
       "      <td>148.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>INCOV147</td>\n",
       "      <td>INCOV147-1</td>\n",
       "      <td>T1</td>\n",
       "      <td>ICU</td>\n",
       "      <td>6</td>\n",
       "      <td>Female</td>\n",
       "      <td>62</td>\n",
       "      <td>NOT Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>36.592466</td>\n",
       "      <td>...</td>\n",
       "      <td>Other</td>\n",
       "      <td>37.7</td>\n",
       "      <td>152.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>139 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Study Subject ID   Sample ID Blood draw time point  \\\n",
       "0           INCOV001  INCOV001-1                    T1   \n",
       "1           INCOV002  INCOV002-1                    T1   \n",
       "2           INCOV003  INCOV003-1                    T1   \n",
       "3           INCOV004  INCOV004-1                    T1   \n",
       "4           INCOV005  INCOV005-1                    T1   \n",
       "..               ...         ...                   ...   \n",
       "134         INCOV141  INCOV141-1                    T1   \n",
       "135         INCOV142  INCOV142-1                    T1   \n",
       "136         INCOV144  INCOV144-1                    T1   \n",
       "137         INCOV145  INCOV145-1                    T1   \n",
       "138         INCOV147  INCOV147-1                    T1   \n",
       "\n",
       "             Patient Location Who Ordinal Scale     Sex  Age  \\\n",
       "0                    Hospital                 3  Female   77   \n",
       "1                         ICU                 5    Male   39   \n",
       "2                         ICU                 7    Male   64   \n",
       "3                    Hospital                 4    Male   76   \n",
       "4                    Hospital                 4    Male   75   \n",
       "..                        ...               ...     ...  ...   \n",
       "134  Home (mobile phlebotomy)                 1    Male   31   \n",
       "135                       ICU                 5    Male   64   \n",
       "136  Home (mobile phlebotomy)            1 or 2  Female   67   \n",
       "137                  Hospital                 4    Male   55   \n",
       "138                       ICU                 6  Female   62   \n",
       "\n",
       "                  Ethnicity   Race        BMI  ...  \\\n",
       "0    NOT Hispanic or Latino  White  33.657783  ...   \n",
       "1    NOT Hispanic or Latino  White  27.900747  ...   \n",
       "2    NOT Hispanic or Latino  White  24.546988  ...   \n",
       "3    NOT Hispanic or Latino  White  23.477620  ...   \n",
       "4    NOT Hispanic or Latino  White  29.703121  ...   \n",
       "..                      ...    ...        ...  ...   \n",
       "134  NOT Hispanic or Latino  White  26.575066  ...   \n",
       "135  NOT Hispanic or Latino  Asian  30.980420  ...   \n",
       "136  NOT Hispanic or Latino  White        NaN  ...   \n",
       "137  NOT Hispanic or Latino  White  35.871500  ...   \n",
       "138  NOT Hispanic or Latino  White  36.592466  ...   \n",
       "\n",
       "                Respiratory Support Temperature Systolic BP Diastolic BP 10X?  \\\n",
       "0                              None        36.6       139.0         60.0  yes   \n",
       "1    High flow nasal cannula (HFNC)        37.3        96.0         57.0  yes   \n",
       "2                             Other        37.2        89.0         50.0  yes   \n",
       "3                     Nasal cannula        36.2       137.0         64.0  yes   \n",
       "4                     Nasal cannula        36.8       126.0        105.0  yes   \n",
       "..                              ...         ...         ...          ...  ...   \n",
       "134                            None         NaN         NaN          NaN  yes   \n",
       "135  High flow nasal cannula (HFNC)        36.2       104.0         45.0  yes   \n",
       "136                            None         NaN         NaN          NaN  yes   \n",
       "137                   Nasal cannula        36.9       148.0         85.0  yes   \n",
       "138                           Other        37.7       152.0         68.0  yes   \n",
       "\n",
       "    Metabolomics Proteomics? Isoplexis CD8 assay? Isoplexis CD4 assay?  \\\n",
       "0            yes         yes                  yes                  yes   \n",
       "1            yes         yes                  yes                  yes   \n",
       "2            yes         yes                  yes                  yes   \n",
       "3            yes         yes                  yes                  yes   \n",
       "4            yes         yes                  yes                  yes   \n",
       "..           ...         ...                  ...                  ...   \n",
       "134           no         yes                   no                   no   \n",
       "135          yes         yes                   no                   no   \n",
       "136           no         yes                   no                   no   \n",
       "137           no         yes                   no                   no   \n",
       "138           no          no                   no                   no   \n",
       "\n",
       "    Isoplexis Monocyte assay?  \n",
       "0                         yes  \n",
       "1                         yes  \n",
       "2                         yes  \n",
       "3                         yes  \n",
       "4                         yes  \n",
       "..                        ...  \n",
       "134                        no  \n",
       "135                        no  \n",
       "136                        no  \n",
       "137                        no  \n",
       "138                        no  \n",
       "\n",
       "[139 rows x 31 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'INCOV001': '3-4', 'INCOV002': '5-7', 'INCOV003': '5-7', 'INCOV004': '3-4', 'INCOV005': '3-4', 'INCOV006': '3-4', 'INCOV007': '3-4', 'INCOV008': '5-7', 'INCOV009': '3-4', 'INCOV010': '3-4', 'INCOV011': '3-4', 'INCOV012': '3-4', 'INCOV013': '5-7', 'INCOV014': '3-4', 'INCOV015': '3-4', 'INCOV016': '3-4', 'INCOV017': '3-4', 'INCOV018': '3-4', 'INCOV019': '3-4', 'INCOV020': '5-7', 'INCOV021': '3-4', 'INCOV022': '5-7', 'INCOV023': '3-4', 'INCOV024': '3-4', 'INCOV025': '5-7', 'INCOV026': '5-7', 'INCOV027': '3-4', 'INCOV028': '3-4', 'INCOV029': '5-7', 'INCOV030': '3-4', 'INCOV031': '5-7', 'INCOV032': '5-7', 'INCOV033': '5-7', 'INCOV034': '3-4', 'INCOV035': '5-7', 'INCOV036': '5-7', 'INCOV037': '3-4', 'INCOV038': '3-4', 'INCOV039': '1-2', 'INCOV040': '3-4', 'INCOV041': '5-7', 'INCOV042': '1-2', 'INCOV043': '3-4', 'INCOV044': '3-4', 'INCOV045': '3-4', 'INCOV047': '5-7', 'INCOV048': '5-7', 'INCOV049': '3-4', 'INCOV050': '3-4', 'INCOV051': '5-7', 'INCOV052': '3-4', 'INCOV053': '3-4', 'INCOV054': '3-4', 'INCOV055': '3-4', 'INCOV056': '1-2', 'INCOV057': '1-2', 'INCOV058': '1-2', 'INCOV059': '3-4', 'INCOV060': '1-2', 'INCOV061': '1-2', 'INCOV062': '3-4', 'INCOV063': '1-2', 'INCOV064': '5-7', 'INCOV066': '1-2', 'INCOV067': '1-2', 'INCOV068': '1-2', 'INCOV069': '3-4', 'INCOV070': '5-7', 'INCOV071': '3-4', 'INCOV072': '3-4', 'INCOV073': '3-4', 'INCOV074': '1-2', 'INCOV075': '5-7', 'INCOV076': '1-2', 'INCOV077': '1-2', 'INCOV078': '1-2', 'INCOV079': '3-4', 'INCOV080': '3-4', 'INCOV081': '1-2', 'INCOV082': '1-2', 'INCOV083': '1-2', 'INCOV084': '5-7', 'INCOV085': '1-2', 'INCOV086': '5-7', 'INCOV087': '5-7', 'INCOV088': '1-2', 'INCOV090': '3-4', 'INCOV091': '3-4', 'INCOV092': '1-2', 'INCOV093': '3-4', 'INCOV094': '1-2', 'INCOV095': '3-4', 'INCOV096': '5-7', 'INCOV097': '1-2', 'INCOV098': '3-4', 'INCOV099': '5-7', 'INCOV100': '1-2', 'INCOV101': '1-2', 'INCOV102': '1-2', 'INCOV103': '1-2', 'INCOV104': '3-4', 'INCOV105': '3-4', 'INCOV106': '5-7', 'INCOV107': '1-2', 'INCOV108': '1-2', 'INCOV109': '1-2', 'INCOV110': '1-2', 'INCOV111': '3-4', 'INCOV112': '1-2', 'INCOV113': '1-2', 'INCOV114': '1-2', 'INCOV115': '3-4', 'INCOV116': '1-2', 'INCOV117': '3-4', 'INCOV119': '1-2', 'INCOV120': '1-2', 'INCOV121': '3-4', 'INCOV122': '1-2', 'INCOV123': '1-2', 'INCOV124': '3-4', 'INCOV125': '3-4', 'INCOV126': '1-2', 'INCOV127': '1-2', 'INCOV128': '1-2', 'INCOV129': '1-2', 'INCOV130': '3-4', 'INCOV131': '1-2', 'INCOV132': '1-2', 'INCOV133': '3-4', 'INCOV134': '1-2', 'INCOV135': '1-2', 'INCOV136': '3-4', 'INCOV137': '5-7', 'INCOV139': '3-4', 'INCOV141': '1-2', 'INCOV142': '5-7', 'INCOV144': '1-2', 'INCOV145': '3-4', 'INCOV147': '5-7'}\n"
     ]
    }
   ],
   "source": [
    "#Read in WHO status data\n",
    "who_data = pd.read_excel('../Data/Su_raw.xlsx', sheet_name=1)\n",
    "\n",
    "#Subset to only first draw readings\n",
    "who_data = who_data[who_data[\"Sample ID\"].str.contains(\"-1\")]\n",
    "who_data = who_data.reset_index(drop=True) #Reset index and drop old index\n",
    "display(who_data)\n",
    "\n",
    "\n",
    "#Create a dictionary with the sample names as a key and WHO status as a value\n",
    "who_dict = {}\n",
    "\n",
    "for i in range(0,len(who_data.index)):\n",
    "    who = who_data[\"Who Ordinal Scale\"][i]\n",
    "    if who == 1 or who == 2 or who == \"1 or 2\":\n",
    "        who = '1-2'\n",
    "    elif who == 3 or who == 4:\n",
    "        who = '3-4'\n",
    "    elif who == 5 or who == 6 or who == 7:\n",
    "        who = '5-7'\n",
    "    else:\n",
    "        print(who)\n",
    "        who = \"ERROR\"\n",
    "\n",
    "    sample = who_data[\"Study Subject ID\"][i]\n",
    "    who_dict[sample] = who\n",
    "\n",
    "\n",
    "\n",
    "print(who_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>WHO_status</th>\n",
       "      <th>Group</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>lectin, galactoside-binding, soluble, 9</th>\n",
       "      <th>tumor necrosis factor receptor superfamily, member 10a</th>\n",
       "      <th>agouti related protein homolog (mouse)</th>\n",
       "      <th>tumor necrosis factor receptor superfamily, member 10b</th>\n",
       "      <th>tumor necrosis factor receptor superfamily, member 13B</th>\n",
       "      <th>c-fos induced growth factor (vascular endothelial growth factor D)</th>\n",
       "      <th>dickkopf WNT signaling pathway inhibitor 1</th>\n",
       "      <th>...</th>\n",
       "      <th>EGF-like-domain, multiple 7</th>\n",
       "      <th>REST corepressor 1</th>\n",
       "      <th>carbonic anhydrase XIV</th>\n",
       "      <th>syntaxin 8</th>\n",
       "      <th>protein kinase, AMP-activated, beta 1 non-catalytic subunit</th>\n",
       "      <th>mitogen-activated protein kinase kinase kinase kinase 5</th>\n",
       "      <th>negative regulator of ubiquitin-like proteins 1</th>\n",
       "      <th>ectonucleoside triphosphate diphosphohydrolase 2</th>\n",
       "      <th>melanoma antigen family D, 1</th>\n",
       "      <th>G protein-coupled receptor 56</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Healthy Donor</td>\n",
       "      <td>1010823</td>\n",
       "      <td>9.467005</td>\n",
       "      <td>3.150612</td>\n",
       "      <td>4.765860</td>\n",
       "      <td>5.853070</td>\n",
       "      <td>9.776093</td>\n",
       "      <td>8.076783</td>\n",
       "      <td>10.286002</td>\n",
       "      <td>...</td>\n",
       "      <td>5.089420</td>\n",
       "      <td>1.904315</td>\n",
       "      <td>2.354585</td>\n",
       "      <td>3.370510</td>\n",
       "      <td>1.457120</td>\n",
       "      <td>6.393335</td>\n",
       "      <td>1.888595</td>\n",
       "      <td>1.272365</td>\n",
       "      <td>-1.015575</td>\n",
       "      <td>1.774740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>Healthy Donor</td>\n",
       "      <td>1027114</td>\n",
       "      <td>2.414755</td>\n",
       "      <td>1.457353</td>\n",
       "      <td>1.244160</td>\n",
       "      <td>0.712265</td>\n",
       "      <td>2.301213</td>\n",
       "      <td>0.910362</td>\n",
       "      <td>-0.073302</td>\n",
       "      <td>...</td>\n",
       "      <td>5.737120</td>\n",
       "      <td>1.706265</td>\n",
       "      <td>2.174810</td>\n",
       "      <td>2.320480</td>\n",
       "      <td>0.754500</td>\n",
       "      <td>3.058775</td>\n",
       "      <td>0.745930</td>\n",
       "      <td>1.413305</td>\n",
       "      <td>-1.683045</td>\n",
       "      <td>1.542735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>Healthy Donor</td>\n",
       "      <td>1029090</td>\n",
       "      <td>9.581175</td>\n",
       "      <td>3.121713</td>\n",
       "      <td>4.649250</td>\n",
       "      <td>5.796290</td>\n",
       "      <td>9.494383</td>\n",
       "      <td>8.618412</td>\n",
       "      <td>9.356512</td>\n",
       "      <td>...</td>\n",
       "      <td>5.046855</td>\n",
       "      <td>1.923825</td>\n",
       "      <td>2.362200</td>\n",
       "      <td>3.534835</td>\n",
       "      <td>1.649450</td>\n",
       "      <td>5.595935</td>\n",
       "      <td>2.099020</td>\n",
       "      <td>1.539440</td>\n",
       "      <td>-0.599360</td>\n",
       "      <td>1.543860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>Healthy Donor</td>\n",
       "      <td>1049423</td>\n",
       "      <td>9.391395</td>\n",
       "      <td>3.068790</td>\n",
       "      <td>4.669920</td>\n",
       "      <td>5.773255</td>\n",
       "      <td>11.887820</td>\n",
       "      <td>7.702655</td>\n",
       "      <td>9.278815</td>\n",
       "      <td>...</td>\n",
       "      <td>4.665445</td>\n",
       "      <td>1.910215</td>\n",
       "      <td>2.364895</td>\n",
       "      <td>3.594625</td>\n",
       "      <td>1.468635</td>\n",
       "      <td>5.781550</td>\n",
       "      <td>1.864265</td>\n",
       "      <td>1.657420</td>\n",
       "      <td>-0.094040</td>\n",
       "      <td>1.229660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>Healthy Donor</td>\n",
       "      <td>1057525</td>\n",
       "      <td>9.142025</td>\n",
       "      <td>3.198663</td>\n",
       "      <td>5.430080</td>\n",
       "      <td>5.614800</td>\n",
       "      <td>9.746692</td>\n",
       "      <td>8.250982</td>\n",
       "      <td>9.407002</td>\n",
       "      <td>...</td>\n",
       "      <td>5.510625</td>\n",
       "      <td>1.770395</td>\n",
       "      <td>2.533960</td>\n",
       "      <td>3.393515</td>\n",
       "      <td>1.293010</td>\n",
       "      <td>5.925985</td>\n",
       "      <td>1.124380</td>\n",
       "      <td>1.383710</td>\n",
       "      <td>-0.680310</td>\n",
       "      <td>1.763090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>3-4</td>\n",
       "      <td>COVID19</td>\n",
       "      <td>INCOV139</td>\n",
       "      <td>9.786717</td>\n",
       "      <td>3.699688</td>\n",
       "      <td>4.876522</td>\n",
       "      <td>6.064768</td>\n",
       "      <td>9.259790</td>\n",
       "      <td>7.528678</td>\n",
       "      <td>11.526053</td>\n",
       "      <td>...</td>\n",
       "      <td>6.731739</td>\n",
       "      <td>5.465761</td>\n",
       "      <td>2.660046</td>\n",
       "      <td>6.218315</td>\n",
       "      <td>4.499703</td>\n",
       "      <td>9.236677</td>\n",
       "      <td>5.852625</td>\n",
       "      <td>1.008050</td>\n",
       "      <td>1.686931</td>\n",
       "      <td>1.697320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>1-2</td>\n",
       "      <td>COVID19</td>\n",
       "      <td>INCOV141</td>\n",
       "      <td>10.382318</td>\n",
       "      <td>3.511383</td>\n",
       "      <td>4.385330</td>\n",
       "      <td>5.207974</td>\n",
       "      <td>9.330641</td>\n",
       "      <td>7.803160</td>\n",
       "      <td>10.952554</td>\n",
       "      <td>...</td>\n",
       "      <td>6.131039</td>\n",
       "      <td>3.616375</td>\n",
       "      <td>3.294628</td>\n",
       "      <td>5.700506</td>\n",
       "      <td>3.362275</td>\n",
       "      <td>8.424453</td>\n",
       "      <td>4.588876</td>\n",
       "      <td>1.053973</td>\n",
       "      <td>0.996480</td>\n",
       "      <td>1.344471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>5-7</td>\n",
       "      <td>COVID19</td>\n",
       "      <td>INCOV142</td>\n",
       "      <td>10.908705</td>\n",
       "      <td>4.361646</td>\n",
       "      <td>5.858950</td>\n",
       "      <td>6.801220</td>\n",
       "      <td>10.706707</td>\n",
       "      <td>6.956506</td>\n",
       "      <td>11.226867</td>\n",
       "      <td>...</td>\n",
       "      <td>6.102758</td>\n",
       "      <td>3.627333</td>\n",
       "      <td>1.468618</td>\n",
       "      <td>4.953478</td>\n",
       "      <td>3.224846</td>\n",
       "      <td>8.259428</td>\n",
       "      <td>3.961976</td>\n",
       "      <td>0.972515</td>\n",
       "      <td>0.772288</td>\n",
       "      <td>3.442974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>1-2</td>\n",
       "      <td>COVID19</td>\n",
       "      <td>INCOV144</td>\n",
       "      <td>9.624015</td>\n",
       "      <td>3.330472</td>\n",
       "      <td>4.729950</td>\n",
       "      <td>4.996068</td>\n",
       "      <td>9.785855</td>\n",
       "      <td>7.849917</td>\n",
       "      <td>10.436921</td>\n",
       "      <td>...</td>\n",
       "      <td>5.427625</td>\n",
       "      <td>2.387093</td>\n",
       "      <td>1.869467</td>\n",
       "      <td>5.393703</td>\n",
       "      <td>3.315091</td>\n",
       "      <td>8.689472</td>\n",
       "      <td>3.920355</td>\n",
       "      <td>1.406586</td>\n",
       "      <td>0.873741</td>\n",
       "      <td>1.418509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>3-4</td>\n",
       "      <td>COVID19</td>\n",
       "      <td>INCOV145</td>\n",
       "      <td>11.174345</td>\n",
       "      <td>4.703725</td>\n",
       "      <td>7.628121</td>\n",
       "      <td>9.758589</td>\n",
       "      <td>10.389508</td>\n",
       "      <td>7.250206</td>\n",
       "      <td>11.577085</td>\n",
       "      <td>...</td>\n",
       "      <td>7.083230</td>\n",
       "      <td>3.670221</td>\n",
       "      <td>2.719823</td>\n",
       "      <td>6.030698</td>\n",
       "      <td>3.917369</td>\n",
       "      <td>9.252264</td>\n",
       "      <td>4.815057</td>\n",
       "      <td>1.007878</td>\n",
       "      <td>1.571172</td>\n",
       "      <td>3.175467</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>259 rows × 467 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    WHO_status           Group sample_id  \\\n",
       "0            0  Healthy Donor    1010823   \n",
       "1            0  Healthy Donor    1027114   \n",
       "2            0  Healthy Donor    1029090   \n",
       "3            0  Healthy Donor    1049423   \n",
       "4            0  Healthy Donor    1057525   \n",
       "..         ...             ...       ...   \n",
       "254        3-4        COVID19   INCOV139   \n",
       "255        1-2        COVID19   INCOV141   \n",
       "256        5-7        COVID19   INCOV142   \n",
       "257        1-2        COVID19   INCOV144   \n",
       "258        3-4        COVID19   INCOV145   \n",
       "\n",
       "     lectin, galactoside-binding, soluble, 9  \\\n",
       "0                                   9.467005   \n",
       "1                                   2.414755   \n",
       "2                                   9.581175   \n",
       "3                                   9.391395   \n",
       "4                                   9.142025   \n",
       "..                                       ...   \n",
       "254                                 9.786717   \n",
       "255                                10.382318   \n",
       "256                                10.908705   \n",
       "257                                 9.624015   \n",
       "258                                11.174345   \n",
       "\n",
       "     tumor necrosis factor receptor superfamily, member 10a  \\\n",
       "0                                             3.150612        \n",
       "1                                             1.457353        \n",
       "2                                             3.121713        \n",
       "3                                             3.068790        \n",
       "4                                             3.198663        \n",
       "..                                                 ...        \n",
       "254                                           3.699688        \n",
       "255                                           3.511383        \n",
       "256                                           4.361646        \n",
       "257                                           3.330472        \n",
       "258                                           4.703725        \n",
       "\n",
       "     agouti related protein homolog (mouse)  \\\n",
       "0                                  4.765860   \n",
       "1                                  1.244160   \n",
       "2                                  4.649250   \n",
       "3                                  4.669920   \n",
       "4                                  5.430080   \n",
       "..                                      ...   \n",
       "254                                4.876522   \n",
       "255                                4.385330   \n",
       "256                                5.858950   \n",
       "257                                4.729950   \n",
       "258                                7.628121   \n",
       "\n",
       "     tumor necrosis factor receptor superfamily, member 10b  \\\n",
       "0                                             5.853070        \n",
       "1                                             0.712265        \n",
       "2                                             5.796290        \n",
       "3                                             5.773255        \n",
       "4                                             5.614800        \n",
       "..                                                 ...        \n",
       "254                                           6.064768        \n",
       "255                                           5.207974        \n",
       "256                                           6.801220        \n",
       "257                                           4.996068        \n",
       "258                                           9.758589        \n",
       "\n",
       "     tumor necrosis factor receptor superfamily, member 13B  \\\n",
       "0                                             9.776093        \n",
       "1                                             2.301213        \n",
       "2                                             9.494383        \n",
       "3                                            11.887820        \n",
       "4                                             9.746692        \n",
       "..                                                 ...        \n",
       "254                                           9.259790        \n",
       "255                                           9.330641        \n",
       "256                                          10.706707        \n",
       "257                                           9.785855        \n",
       "258                                          10.389508        \n",
       "\n",
       "     c-fos induced growth factor (vascular endothelial growth factor D)  \\\n",
       "0                                             8.076783                    \n",
       "1                                             0.910362                    \n",
       "2                                             8.618412                    \n",
       "3                                             7.702655                    \n",
       "4                                             8.250982                    \n",
       "..                                                 ...                    \n",
       "254                                           7.528678                    \n",
       "255                                           7.803160                    \n",
       "256                                           6.956506                    \n",
       "257                                           7.849917                    \n",
       "258                                           7.250206                    \n",
       "\n",
       "     dickkopf WNT signaling pathway inhibitor 1  ...  \\\n",
       "0                                     10.286002  ...   \n",
       "1                                     -0.073302  ...   \n",
       "2                                      9.356512  ...   \n",
       "3                                      9.278815  ...   \n",
       "4                                      9.407002  ...   \n",
       "..                                          ...  ...   \n",
       "254                                   11.526053  ...   \n",
       "255                                   10.952554  ...   \n",
       "256                                   11.226867  ...   \n",
       "257                                   10.436921  ...   \n",
       "258                                   11.577085  ...   \n",
       "\n",
       "     EGF-like-domain, multiple 7  REST corepressor 1  carbonic anhydrase XIV  \\\n",
       "0                       5.089420            1.904315                2.354585   \n",
       "1                       5.737120            1.706265                2.174810   \n",
       "2                       5.046855            1.923825                2.362200   \n",
       "3                       4.665445            1.910215                2.364895   \n",
       "4                       5.510625            1.770395                2.533960   \n",
       "..                           ...                 ...                     ...   \n",
       "254                     6.731739            5.465761                2.660046   \n",
       "255                     6.131039            3.616375                3.294628   \n",
       "256                     6.102758            3.627333                1.468618   \n",
       "257                     5.427625            2.387093                1.869467   \n",
       "258                     7.083230            3.670221                2.719823   \n",
       "\n",
       "     syntaxin 8  protein kinase, AMP-activated, beta 1 non-catalytic subunit  \\\n",
       "0      3.370510                                           1.457120             \n",
       "1      2.320480                                           0.754500             \n",
       "2      3.534835                                           1.649450             \n",
       "3      3.594625                                           1.468635             \n",
       "4      3.393515                                           1.293010             \n",
       "..          ...                                                ...             \n",
       "254    6.218315                                           4.499703             \n",
       "255    5.700506                                           3.362275             \n",
       "256    4.953478                                           3.224846             \n",
       "257    5.393703                                           3.315091             \n",
       "258    6.030698                                           3.917369             \n",
       "\n",
       "     mitogen-activated protein kinase kinase kinase kinase 5  \\\n",
       "0                                             6.393335         \n",
       "1                                             3.058775         \n",
       "2                                             5.595935         \n",
       "3                                             5.781550         \n",
       "4                                             5.925985         \n",
       "..                                                 ...         \n",
       "254                                           9.236677         \n",
       "255                                           8.424453         \n",
       "256                                           8.259428         \n",
       "257                                           8.689472         \n",
       "258                                           9.252264         \n",
       "\n",
       "     negative regulator of ubiquitin-like proteins 1  \\\n",
       "0                                           1.888595   \n",
       "1                                           0.745930   \n",
       "2                                           2.099020   \n",
       "3                                           1.864265   \n",
       "4                                           1.124380   \n",
       "..                                               ...   \n",
       "254                                         5.852625   \n",
       "255                                         4.588876   \n",
       "256                                         3.961976   \n",
       "257                                         3.920355   \n",
       "258                                         4.815057   \n",
       "\n",
       "     ectonucleoside triphosphate diphosphohydrolase 2  \\\n",
       "0                                            1.272365   \n",
       "1                                            1.413305   \n",
       "2                                            1.539440   \n",
       "3                                            1.657420   \n",
       "4                                            1.383710   \n",
       "..                                                ...   \n",
       "254                                          1.008050   \n",
       "255                                          1.053973   \n",
       "256                                          0.972515   \n",
       "257                                          1.406586   \n",
       "258                                          1.007878   \n",
       "\n",
       "     melanoma antigen family D, 1  G protein-coupled receptor 56  \n",
       "0                       -1.015575                       1.774740  \n",
       "1                       -1.683045                       1.542735  \n",
       "2                       -0.599360                       1.543860  \n",
       "3                       -0.094040                       1.229660  \n",
       "4                       -0.680310                       1.763090  \n",
       "..                            ...                            ...  \n",
       "254                      1.686931                       1.697320  \n",
       "255                      0.996480                       1.344471  \n",
       "256                      0.772288                       3.442974  \n",
       "257                      0.873741                       1.418509  \n",
       "258                      1.571172                       3.175467  \n",
       "\n",
       "[259 rows x 467 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Add WHO data into the original dataframe\n",
    "for i in range(124,len(df.index)): #first instance of a case is at index 124\n",
    "    sample = df.loc[i,\"sample_id\"]\n",
    "    if sample in who_dict:\n",
    "        df.loc[i,\"WHO_status\"] = who_dict[sample]\n",
    "\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>WHO_status</th>\n",
       "      <th>Group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1010823</th>\n",
       "      <td>0</td>\n",
       "      <td>Healthy Donor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1027114</th>\n",
       "      <td>0</td>\n",
       "      <td>Healthy Donor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1029090</th>\n",
       "      <td>0</td>\n",
       "      <td>Healthy Donor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1049423</th>\n",
       "      <td>0</td>\n",
       "      <td>Healthy Donor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1057525</th>\n",
       "      <td>0</td>\n",
       "      <td>Healthy Donor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV139</th>\n",
       "      <td>3-4</td>\n",
       "      <td>COVID19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV141</th>\n",
       "      <td>1-2</td>\n",
       "      <td>COVID19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV142</th>\n",
       "      <td>5-7</td>\n",
       "      <td>COVID19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV144</th>\n",
       "      <td>1-2</td>\n",
       "      <td>COVID19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV145</th>\n",
       "      <td>3-4</td>\n",
       "      <td>COVID19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>259 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          WHO_status           Group\n",
       "sample_id                           \n",
       "1010823            0  Healthy Donor \n",
       "1027114            0  Healthy Donor \n",
       "1029090            0  Healthy Donor \n",
       "1049423            0  Healthy Donor \n",
       "1057525            0  Healthy Donor \n",
       "...              ...             ...\n",
       "INCOV139         3-4        COVID19 \n",
       "INCOV141         1-2        COVID19 \n",
       "INCOV142         5-7        COVID19 \n",
       "INCOV144         1-2        COVID19 \n",
       "INCOV145         3-4        COVID19 \n",
       "\n",
       "[259 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Make a different dataframe containing the meta data only (this will be changed later when samples are removed)\n",
    "df.set_index('sample_id', inplace=True)\n",
    "df.columns.values.tolist()\n",
    "\n",
    "meta_data = df.iloc[:,[0,1]]\n",
    "display(meta_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.5502263347090932\n"
     ]
    }
   ],
   "source": [
    "df_num  = df.iloc[:,2:] #all rows, all columns apart from the first three which is meta data\n",
    "#df_num = df_num.set_index('sample_id')\n",
    "\n",
    "#Percentage of missing data in dataframe\n",
    "missing_vals = df_num.isna().sum().sum()\n",
    "print((missing_vals/(259*464))*100) #1.6% missing data for proteomic\n",
    "#print((missing_vals/(266*1050))*100) #19.3% for metabolomic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "interleukin 24                                             True\n",
      "sprouty homolog 2 (Drosophila)                             True\n",
      "protein phosphatase 1, regulatory (inhibitor) subunit 2    True\n",
      "clusterin-like 1 (retinal)                                 True\n",
      "dtype: bool\n"
     ]
    }
   ],
   "source": [
    "#Identify which columns have missing data\n",
    "#print(df_num.isna().any()[lambda x: x])\n",
    "\n",
    "#Remove features with >40% missing data - same as Sara's thesis\n",
    "df_filter = df_num.loc[:, df_num.isin([' ', np.nan, 0]).mean() <= 0.4]    #keep all rows, if columns have ' ' or nan or 0, count as though missing and take as TRUE, get the mean of all the TRUE values\n",
    "print(df_filter.isna().any()[lambda x: x]) #Identify which columns still have missing data\n",
    "\n",
    "\n",
    "#For metabolomic 895 metabolites left after filtering\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1029090 clusterin-like 1 (retinal)\n",
      "1057525 clusterin-like 1 (retinal)\n",
      "1097616 clusterin-like 1 (retinal)\n",
      "1100093 clusterin-like 1 (retinal)\n",
      "1129657 clusterin-like 1 (retinal)\n",
      "1157578 sprouty homolog 2 (Drosophila)\n",
      "1221615 clusterin-like 1 (retinal)\n",
      "1268568 clusterin-like 1 (retinal)\n",
      "1305376 clusterin-like 1 (retinal)\n",
      "1310874 clusterin-like 1 (retinal)\n",
      "1319761 clusterin-like 1 (retinal)\n",
      "1358815 clusterin-like 1 (retinal)\n",
      "1391396 clusterin-like 1 (retinal)\n",
      "1420546 clusterin-like 1 (retinal)\n",
      "1438254 clusterin-like 1 (retinal)\n",
      "1532234 clusterin-like 1 (retinal)\n",
      "1541442 clusterin-like 1 (retinal)\n",
      "1550067 clusterin-like 1 (retinal)\n",
      "1673724 clusterin-like 1 (retinal)\n",
      "1782796 clusterin-like 1 (retinal)\n",
      "1794557 clusterin-like 1 (retinal)\n",
      "1824736 interleukin 24\n",
      "1881570 clusterin-like 1 (retinal)\n",
      "1916238 clusterin-like 1 (retinal)\n",
      "1927208 clusterin-like 1 (retinal)\n",
      "1975030 clusterin-like 1 (retinal)\n",
      "INCOV034 interleukin 24\n",
      "INCOV107 protein phosphatase 1, regulatory (inhibitor) subunit 2\n",
      "0.02381235861412073\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lectin, galactoside-binding, soluble, 9</th>\n",
       "      <th>tumor necrosis factor receptor superfamily, member 10a</th>\n",
       "      <th>agouti related protein homolog (mouse)</th>\n",
       "      <th>tumor necrosis factor receptor superfamily, member 10b</th>\n",
       "      <th>tumor necrosis factor receptor superfamily, member 13B</th>\n",
       "      <th>c-fos induced growth factor (vascular endothelial growth factor D)</th>\n",
       "      <th>dickkopf WNT signaling pathway inhibitor 1</th>\n",
       "      <th>renin</th>\n",
       "      <th>platelet-derived growth factor beta polypeptide</th>\n",
       "      <th>growth hormone 1</th>\n",
       "      <th>...</th>\n",
       "      <th>EGF-like-domain, multiple 7</th>\n",
       "      <th>REST corepressor 1</th>\n",
       "      <th>carbonic anhydrase XIV</th>\n",
       "      <th>syntaxin 8</th>\n",
       "      <th>protein kinase, AMP-activated, beta 1 non-catalytic subunit</th>\n",
       "      <th>mitogen-activated protein kinase kinase kinase kinase 5</th>\n",
       "      <th>negative regulator of ubiquitin-like proteins 1</th>\n",
       "      <th>ectonucleoside triphosphate diphosphohydrolase 2</th>\n",
       "      <th>melanoma antigen family D, 1</th>\n",
       "      <th>G protein-coupled receptor 56</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1010823</th>\n",
       "      <td>9.467005</td>\n",
       "      <td>3.150612</td>\n",
       "      <td>4.765860</td>\n",
       "      <td>5.853070</td>\n",
       "      <td>9.776093</td>\n",
       "      <td>8.076783</td>\n",
       "      <td>10.286002</td>\n",
       "      <td>7.445927</td>\n",
       "      <td>11.845645</td>\n",
       "      <td>10.276995</td>\n",
       "      <td>...</td>\n",
       "      <td>5.089420</td>\n",
       "      <td>1.904315</td>\n",
       "      <td>2.354585</td>\n",
       "      <td>3.370510</td>\n",
       "      <td>1.457120</td>\n",
       "      <td>6.393335</td>\n",
       "      <td>1.888595</td>\n",
       "      <td>1.272365</td>\n",
       "      <td>-1.015575</td>\n",
       "      <td>1.774740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1027114</th>\n",
       "      <td>2.414755</td>\n",
       "      <td>1.457353</td>\n",
       "      <td>1.244160</td>\n",
       "      <td>0.712265</td>\n",
       "      <td>2.301213</td>\n",
       "      <td>0.910362</td>\n",
       "      <td>-0.073302</td>\n",
       "      <td>1.053592</td>\n",
       "      <td>0.319995</td>\n",
       "      <td>0.783655</td>\n",
       "      <td>...</td>\n",
       "      <td>5.737120</td>\n",
       "      <td>1.706265</td>\n",
       "      <td>2.174810</td>\n",
       "      <td>2.320480</td>\n",
       "      <td>0.754500</td>\n",
       "      <td>3.058775</td>\n",
       "      <td>0.745930</td>\n",
       "      <td>1.413305</td>\n",
       "      <td>-1.683045</td>\n",
       "      <td>1.542735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1029090</th>\n",
       "      <td>9.581175</td>\n",
       "      <td>3.121713</td>\n",
       "      <td>4.649250</td>\n",
       "      <td>5.796290</td>\n",
       "      <td>9.494383</td>\n",
       "      <td>8.618412</td>\n",
       "      <td>9.356512</td>\n",
       "      <td>7.023788</td>\n",
       "      <td>10.991955</td>\n",
       "      <td>9.868815</td>\n",
       "      <td>...</td>\n",
       "      <td>5.046855</td>\n",
       "      <td>1.923825</td>\n",
       "      <td>2.362200</td>\n",
       "      <td>3.534835</td>\n",
       "      <td>1.649450</td>\n",
       "      <td>5.595935</td>\n",
       "      <td>2.099020</td>\n",
       "      <td>1.539440</td>\n",
       "      <td>-0.599360</td>\n",
       "      <td>1.543860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1049423</th>\n",
       "      <td>9.391395</td>\n",
       "      <td>3.068790</td>\n",
       "      <td>4.669920</td>\n",
       "      <td>5.773255</td>\n",
       "      <td>11.887820</td>\n",
       "      <td>7.702655</td>\n",
       "      <td>9.278815</td>\n",
       "      <td>8.153380</td>\n",
       "      <td>10.270445</td>\n",
       "      <td>5.840475</td>\n",
       "      <td>...</td>\n",
       "      <td>4.665445</td>\n",
       "      <td>1.910215</td>\n",
       "      <td>2.364895</td>\n",
       "      <td>3.594625</td>\n",
       "      <td>1.468635</td>\n",
       "      <td>5.781550</td>\n",
       "      <td>1.864265</td>\n",
       "      <td>1.657420</td>\n",
       "      <td>-0.094040</td>\n",
       "      <td>1.229660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1057525</th>\n",
       "      <td>9.142025</td>\n",
       "      <td>3.198663</td>\n",
       "      <td>5.430080</td>\n",
       "      <td>5.614800</td>\n",
       "      <td>9.746692</td>\n",
       "      <td>8.250982</td>\n",
       "      <td>9.407002</td>\n",
       "      <td>7.645927</td>\n",
       "      <td>11.169755</td>\n",
       "      <td>8.247015</td>\n",
       "      <td>...</td>\n",
       "      <td>5.510625</td>\n",
       "      <td>1.770395</td>\n",
       "      <td>2.533960</td>\n",
       "      <td>3.393515</td>\n",
       "      <td>1.293010</td>\n",
       "      <td>5.925985</td>\n",
       "      <td>1.124380</td>\n",
       "      <td>1.383710</td>\n",
       "      <td>-0.680310</td>\n",
       "      <td>1.763090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV139</th>\n",
       "      <td>9.786717</td>\n",
       "      <td>3.699688</td>\n",
       "      <td>4.876522</td>\n",
       "      <td>6.064768</td>\n",
       "      <td>9.259790</td>\n",
       "      <td>7.528678</td>\n",
       "      <td>11.526053</td>\n",
       "      <td>7.279940</td>\n",
       "      <td>12.224605</td>\n",
       "      <td>8.855252</td>\n",
       "      <td>...</td>\n",
       "      <td>6.731739</td>\n",
       "      <td>5.465761</td>\n",
       "      <td>2.660046</td>\n",
       "      <td>6.218315</td>\n",
       "      <td>4.499703</td>\n",
       "      <td>9.236677</td>\n",
       "      <td>5.852625</td>\n",
       "      <td>1.008050</td>\n",
       "      <td>1.686931</td>\n",
       "      <td>1.697320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV141</th>\n",
       "      <td>10.382318</td>\n",
       "      <td>3.511383</td>\n",
       "      <td>4.385330</td>\n",
       "      <td>5.207974</td>\n",
       "      <td>9.330641</td>\n",
       "      <td>7.803160</td>\n",
       "      <td>10.952554</td>\n",
       "      <td>8.125226</td>\n",
       "      <td>12.180945</td>\n",
       "      <td>7.589490</td>\n",
       "      <td>...</td>\n",
       "      <td>6.131039</td>\n",
       "      <td>3.616375</td>\n",
       "      <td>3.294628</td>\n",
       "      <td>5.700506</td>\n",
       "      <td>3.362275</td>\n",
       "      <td>8.424453</td>\n",
       "      <td>4.588876</td>\n",
       "      <td>1.053973</td>\n",
       "      <td>0.996480</td>\n",
       "      <td>1.344471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV142</th>\n",
       "      <td>10.908705</td>\n",
       "      <td>4.361646</td>\n",
       "      <td>5.858950</td>\n",
       "      <td>6.801220</td>\n",
       "      <td>10.706707</td>\n",
       "      <td>6.956506</td>\n",
       "      <td>11.226867</td>\n",
       "      <td>9.191583</td>\n",
       "      <td>11.765794</td>\n",
       "      <td>4.034886</td>\n",
       "      <td>...</td>\n",
       "      <td>6.102758</td>\n",
       "      <td>3.627333</td>\n",
       "      <td>1.468618</td>\n",
       "      <td>4.953478</td>\n",
       "      <td>3.224846</td>\n",
       "      <td>8.259428</td>\n",
       "      <td>3.961976</td>\n",
       "      <td>0.972515</td>\n",
       "      <td>0.772288</td>\n",
       "      <td>3.442974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV144</th>\n",
       "      <td>9.624015</td>\n",
       "      <td>3.330472</td>\n",
       "      <td>4.729950</td>\n",
       "      <td>4.996068</td>\n",
       "      <td>9.785855</td>\n",
       "      <td>7.849917</td>\n",
       "      <td>10.436921</td>\n",
       "      <td>7.818691</td>\n",
       "      <td>11.536490</td>\n",
       "      <td>7.558431</td>\n",
       "      <td>...</td>\n",
       "      <td>5.427625</td>\n",
       "      <td>2.387093</td>\n",
       "      <td>1.869467</td>\n",
       "      <td>5.393703</td>\n",
       "      <td>3.315091</td>\n",
       "      <td>8.689472</td>\n",
       "      <td>3.920355</td>\n",
       "      <td>1.406586</td>\n",
       "      <td>0.873741</td>\n",
       "      <td>1.418509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INCOV145</th>\n",
       "      <td>11.174345</td>\n",
       "      <td>4.703725</td>\n",
       "      <td>7.628121</td>\n",
       "      <td>9.758589</td>\n",
       "      <td>10.389508</td>\n",
       "      <td>7.250206</td>\n",
       "      <td>11.577085</td>\n",
       "      <td>9.109854</td>\n",
       "      <td>12.079503</td>\n",
       "      <td>5.333670</td>\n",
       "      <td>...</td>\n",
       "      <td>7.083230</td>\n",
       "      <td>3.670221</td>\n",
       "      <td>2.719823</td>\n",
       "      <td>6.030698</td>\n",
       "      <td>3.917369</td>\n",
       "      <td>9.252264</td>\n",
       "      <td>4.815057</td>\n",
       "      <td>1.007878</td>\n",
       "      <td>1.571172</td>\n",
       "      <td>3.175467</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>259 rows × 454 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           lectin, galactoside-binding, soluble, 9  \\\n",
       "sample_id                                            \n",
       "1010823                                   9.467005   \n",
       "1027114                                   2.414755   \n",
       "1029090                                   9.581175   \n",
       "1049423                                   9.391395   \n",
       "1057525                                   9.142025   \n",
       "...                                            ...   \n",
       "INCOV139                                  9.786717   \n",
       "INCOV141                                 10.382318   \n",
       "INCOV142                                 10.908705   \n",
       "INCOV144                                  9.624015   \n",
       "INCOV145                                 11.174345   \n",
       "\n",
       "           tumor necrosis factor receptor superfamily, member 10a  \\\n",
       "sample_id                                                           \n",
       "1010823                                             3.150612        \n",
       "1027114                                             1.457353        \n",
       "1029090                                             3.121713        \n",
       "1049423                                             3.068790        \n",
       "1057525                                             3.198663        \n",
       "...                                                      ...        \n",
       "INCOV139                                            3.699688        \n",
       "INCOV141                                            3.511383        \n",
       "INCOV142                                            4.361646        \n",
       "INCOV144                                            3.330472        \n",
       "INCOV145                                            4.703725        \n",
       "\n",
       "           agouti related protein homolog (mouse)  \\\n",
       "sample_id                                           \n",
       "1010823                                  4.765860   \n",
       "1027114                                  1.244160   \n",
       "1029090                                  4.649250   \n",
       "1049423                                  4.669920   \n",
       "1057525                                  5.430080   \n",
       "...                                           ...   \n",
       "INCOV139                                 4.876522   \n",
       "INCOV141                                 4.385330   \n",
       "INCOV142                                 5.858950   \n",
       "INCOV144                                 4.729950   \n",
       "INCOV145                                 7.628121   \n",
       "\n",
       "           tumor necrosis factor receptor superfamily, member 10b  \\\n",
       "sample_id                                                           \n",
       "1010823                                             5.853070        \n",
       "1027114                                             0.712265        \n",
       "1029090                                             5.796290        \n",
       "1049423                                             5.773255        \n",
       "1057525                                             5.614800        \n",
       "...                                                      ...        \n",
       "INCOV139                                            6.064768        \n",
       "INCOV141                                            5.207974        \n",
       "INCOV142                                            6.801220        \n",
       "INCOV144                                            4.996068        \n",
       "INCOV145                                            9.758589        \n",
       "\n",
       "           tumor necrosis factor receptor superfamily, member 13B  \\\n",
       "sample_id                                                           \n",
       "1010823                                             9.776093        \n",
       "1027114                                             2.301213        \n",
       "1029090                                             9.494383        \n",
       "1049423                                            11.887820        \n",
       "1057525                                             9.746692        \n",
       "...                                                      ...        \n",
       "INCOV139                                            9.259790        \n",
       "INCOV141                                            9.330641        \n",
       "INCOV142                                           10.706707        \n",
       "INCOV144                                            9.785855        \n",
       "INCOV145                                           10.389508        \n",
       "\n",
       "           c-fos induced growth factor (vascular endothelial growth factor D)  \\\n",
       "sample_id                                                                       \n",
       "1010823                                             8.076783                    \n",
       "1027114                                             0.910362                    \n",
       "1029090                                             8.618412                    \n",
       "1049423                                             7.702655                    \n",
       "1057525                                             8.250982                    \n",
       "...                                                      ...                    \n",
       "INCOV139                                            7.528678                    \n",
       "INCOV141                                            7.803160                    \n",
       "INCOV142                                            6.956506                    \n",
       "INCOV144                                            7.849917                    \n",
       "INCOV145                                            7.250206                    \n",
       "\n",
       "           dickkopf WNT signaling pathway inhibitor 1     renin  \\\n",
       "sample_id                                                         \n",
       "1010823                                     10.286002  7.445927   \n",
       "1027114                                     -0.073302  1.053592   \n",
       "1029090                                      9.356512  7.023788   \n",
       "1049423                                      9.278815  8.153380   \n",
       "1057525                                      9.407002  7.645927   \n",
       "...                                               ...       ...   \n",
       "INCOV139                                    11.526053  7.279940   \n",
       "INCOV141                                    10.952554  8.125226   \n",
       "INCOV142                                    11.226867  9.191583   \n",
       "INCOV144                                    10.436921  7.818691   \n",
       "INCOV145                                    11.577085  9.109854   \n",
       "\n",
       "           platelet-derived growth factor beta polypeptide  growth hormone 1  \\\n",
       "sample_id                                                                      \n",
       "1010823                                          11.845645         10.276995   \n",
       "1027114                                           0.319995          0.783655   \n",
       "1029090                                          10.991955          9.868815   \n",
       "1049423                                          10.270445          5.840475   \n",
       "1057525                                          11.169755          8.247015   \n",
       "...                                                    ...               ...   \n",
       "INCOV139                                         12.224605          8.855252   \n",
       "INCOV141                                         12.180945          7.589490   \n",
       "INCOV142                                         11.765794          4.034886   \n",
       "INCOV144                                         11.536490          7.558431   \n",
       "INCOV145                                         12.079503          5.333670   \n",
       "\n",
       "           ...  EGF-like-domain, multiple 7  REST corepressor 1  \\\n",
       "sample_id  ...                                                    \n",
       "1010823    ...                     5.089420            1.904315   \n",
       "1027114    ...                     5.737120            1.706265   \n",
       "1029090    ...                     5.046855            1.923825   \n",
       "1049423    ...                     4.665445            1.910215   \n",
       "1057525    ...                     5.510625            1.770395   \n",
       "...        ...                          ...                 ...   \n",
       "INCOV139   ...                     6.731739            5.465761   \n",
       "INCOV141   ...                     6.131039            3.616375   \n",
       "INCOV142   ...                     6.102758            3.627333   \n",
       "INCOV144   ...                     5.427625            2.387093   \n",
       "INCOV145   ...                     7.083230            3.670221   \n",
       "\n",
       "           carbonic anhydrase XIV  syntaxin 8  \\\n",
       "sample_id                                       \n",
       "1010823                  2.354585    3.370510   \n",
       "1027114                  2.174810    2.320480   \n",
       "1029090                  2.362200    3.534835   \n",
       "1049423                  2.364895    3.594625   \n",
       "1057525                  2.533960    3.393515   \n",
       "...                           ...         ...   \n",
       "INCOV139                 2.660046    6.218315   \n",
       "INCOV141                 3.294628    5.700506   \n",
       "INCOV142                 1.468618    4.953478   \n",
       "INCOV144                 1.869467    5.393703   \n",
       "INCOV145                 2.719823    6.030698   \n",
       "\n",
       "           protein kinase, AMP-activated, beta 1 non-catalytic subunit  \\\n",
       "sample_id                                                                \n",
       "1010823                                             1.457120             \n",
       "1027114                                             0.754500             \n",
       "1029090                                             1.649450             \n",
       "1049423                                             1.468635             \n",
       "1057525                                             1.293010             \n",
       "...                                                      ...             \n",
       "INCOV139                                            4.499703             \n",
       "INCOV141                                            3.362275             \n",
       "INCOV142                                            3.224846             \n",
       "INCOV144                                            3.315091             \n",
       "INCOV145                                            3.917369             \n",
       "\n",
       "           mitogen-activated protein kinase kinase kinase kinase 5  \\\n",
       "sample_id                                                            \n",
       "1010823                                             6.393335         \n",
       "1027114                                             3.058775         \n",
       "1029090                                             5.595935         \n",
       "1049423                                             5.781550         \n",
       "1057525                                             5.925985         \n",
       "...                                                      ...         \n",
       "INCOV139                                            9.236677         \n",
       "INCOV141                                            8.424453         \n",
       "INCOV142                                            8.259428         \n",
       "INCOV144                                            8.689472         \n",
       "INCOV145                                            9.252264         \n",
       "\n",
       "           negative regulator of ubiquitin-like proteins 1  \\\n",
       "sample_id                                                    \n",
       "1010823                                           1.888595   \n",
       "1027114                                           0.745930   \n",
       "1029090                                           2.099020   \n",
       "1049423                                           1.864265   \n",
       "1057525                                           1.124380   \n",
       "...                                                    ...   \n",
       "INCOV139                                          5.852625   \n",
       "INCOV141                                          4.588876   \n",
       "INCOV142                                          3.961976   \n",
       "INCOV144                                          3.920355   \n",
       "INCOV145                                          4.815057   \n",
       "\n",
       "           ectonucleoside triphosphate diphosphohydrolase 2  \\\n",
       "sample_id                                                     \n",
       "1010823                                            1.272365   \n",
       "1027114                                            1.413305   \n",
       "1029090                                            1.539440   \n",
       "1049423                                            1.657420   \n",
       "1057525                                            1.383710   \n",
       "...                                                     ...   \n",
       "INCOV139                                           1.008050   \n",
       "INCOV141                                           1.053973   \n",
       "INCOV142                                           0.972515   \n",
       "INCOV144                                           1.406586   \n",
       "INCOV145                                           1.007878   \n",
       "\n",
       "           melanoma antigen family D, 1  G protein-coupled receptor 56  \n",
       "sample_id                                                               \n",
       "1010823                       -1.015575                       1.774740  \n",
       "1027114                       -1.683045                       1.542735  \n",
       "1029090                       -0.599360                       1.543860  \n",
       "1049423                       -0.094040                       1.229660  \n",
       "1057525                       -0.680310                       1.763090  \n",
       "...                                 ...                            ...  \n",
       "INCOV139                       1.686931                       1.697320  \n",
       "INCOV141                       0.996480                       1.344471  \n",
       "INCOV142                       0.772288                       3.442974  \n",
       "INCOV144                       0.873741                       1.418509  \n",
       "INCOV145                       1.571172                       3.175467  \n",
       "\n",
       "[259 rows x 454 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "#Identify which rows have missing data\n",
    "#print(df_filter[df_filter.isna().any(axis=1)])\n",
    "\n",
    "#identify which ROWS have missing values for each sample\n",
    "df_missing_row = df_filter[df_filter.isna().any(axis=1)]\n",
    "for index, row in df_missing_row.iterrows():\n",
    "    for column in df_missing_row.columns:\n",
    "        if pd.isnull(row[column]):\n",
    "            print(index,column)\n",
    "\n",
    "#Remove samples with >50% missing data - same as Sara's thesis (none of the samples were removed)\n",
    "df_filter = df_filter.loc[df_filter.isin([' ', np.nan, 0]).mean(axis=1) <= 0.5,:]    \n",
    "\n",
    "missing_vals = df_filter.isna().sum().sum()    #.sum() takes sum of missing values in each column, so .sum() again to take sum of all column values\n",
    "print((missing_vals/(259*454))*100)  #259 rows, 454 proteins\n",
    "\n",
    "display(df_filter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'MissForest' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[11], line 2\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[39m#Impute missing data (ignore the warning messages)\u001b[39;00m\n\u001b[1;32m----> 2\u001b[0m imputer \u001b[39m=\u001b[39m MissForest()\n\u001b[0;32m      3\u001b[0m X_imputed \u001b[39m=\u001b[39m imputer\u001b[39m.\u001b[39mfit_transform(df_filter\u001b[39m.\u001b[39mto_numpy())\n",
      "\u001b[1;31mNameError\u001b[0m: name 'MissForest' is not defined"
     ]
    }
   ],
   "source": [
    "#Impute missing data (ignore the warning messages)\n",
    "imputer = MissForest()\n",
    "X_imputed = imputer.fit_transform(df_filter.to_numpy())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Check imputed data to make sure there are no missing values\n",
    "df_impute = pd.DataFrame(X_imputed,columns=df_filter.columns, index=df_filter.index)\n",
    "missing_vals = df_impute.isna().sum().sum()\n",
    "print((missing_vals/(259*464))*100) #no missing values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#The data has already gone through log2normalisation (negative values present frmo values < 1)\n",
    "#Taking log2 of a number under 1 will give a negative value\n",
    "#Note: log2 transforming is a normalisation step, which usually reduces the skew in the data, so that the data can be scaled using StandardScaler()\n",
    "print(df_impute.max().max())\n",
    "print(df_impute.min().min())\n",
    "print(df_impute.mean(axis = 0)) #if scaled -> mean of 0   (not scaled yet)\n",
    "print(df_impute.std(axis = 0)) #if scaled -> sd of 1       (not scaled yet)\n",
    "\n",
    "display(df_impute)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_hist = df_impute.mean(axis = 0)\n",
    "sns.histplot(df_hist, bins = 30,color='#BABAE2',edgecolor=\"k\") \n",
    "\n",
    "#The mean value for each metabolite has been plotted\n",
    "plt.title('Mean protein distribution before scaling',fontsize=16)\n",
    "plt.xlabel('Protein expression',fontsize=13)\n",
    "plt.ylabel('Count',fontsize=13) \n",
    "\n",
    "#plt.savefig( 'Figures/prescale_mean_protein_distribution.png' , dpi=200,bbox_inches = 'tight' , pad_inches = 0.2 , facecolor='w')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_impute_np = df_impute.to_numpy()\n",
    "df_hist = df_impute_np.flatten()\n",
    "sns.histplot(df_hist, bins = 30,color='#BABAE2',edgecolor=\"k\") \n",
    "\n",
    "#The mean value for each metabolite has been plotted\n",
    "plt.title('Protein distribution before scaling',fontsize=16)\n",
    "plt.xlabel('Protein expression',fontsize=13)\n",
    "plt.ylabel('Count',fontsize=13) \n",
    "\n",
    "#plt.savefig( 'Figures/prescale_protein_distribution.png' , dpi=200,bbox_inches = 'tight' , pad_inches = 0.2 , facecolor='w')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#I decided not to use because too stringent (>20 outliers detected)\n",
    "\n",
    "# Carry out PCA to identify outliers  \n",
    "#https://towardsdatascience.com/outlier-detection-using-principal-component-analysis-and-hotellings-t2-and-spe-dmodx-methods-625b3c90897\n",
    "\n",
    "#from pca import pca\n",
    "\n",
    "# Initialize pca to also detected outliers\n",
    "#model = pca(normalize=True, detect_outliers=['ht2', 'spe'], n_std=2  )\n",
    "\n",
    "# Fit and transform\n",
    "#results = model.fit_transform(df_impute)\n",
    "\n",
    "# Get the outliers using Hotellings T2 method\n",
    "#df_impute.loc[results['outliers']['y_bool'], :]\n",
    "\n",
    "# Get the outliers using SPE/DmodX method\n",
    "#df_impute.loc[results['outliers']['y_bool_spe'], :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df_impute' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[12], line 3\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[39m#Conduct PCA to identify outliers\u001b[39;00m\n\u001b[1;32m----> 3\u001b[0m features \u001b[39m=\u001b[39m df_impute\u001b[39m.\u001b[39mcolumns[:]\n\u001b[0;32m      4\u001b[0m x \u001b[39m=\u001b[39m df_impute\u001b[39m.\u001b[39mloc[:, features]\u001b[39m.\u001b[39mvalues\n\u001b[0;32m      6\u001b[0m pca \u001b[39m=\u001b[39m PCA(n_components\u001b[39m=\u001b[39m\u001b[39m20\u001b[39m) \n",
      "\u001b[1;31mNameError\u001b[0m: name 'df_impute' is not defined"
     ]
    }
   ],
   "source": [
    "#Conduct PCA to identify outliers\n",
    "\n",
    "features = df_impute.columns[:]\n",
    "x = df_impute.loc[:, features].values\n",
    "\n",
    "pca = PCA(n_components=20) \n",
    "principal_components = pca.fit_transform(x)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Use elbow method for scree plot to determine number of PCs to use to calculate residuals (4 principal components)\n",
    "print(sum(pca.explained_variance_ratio_))\n",
    "PC_values = np.arange(pca.n_components_) + 1\n",
    "plt.plot(PC_values, pca.explained_variance_ratio_, 'o-', linewidth=2, color='blue')\n",
    "\n",
    "\n",
    "sns.set_style(\"white\")\n",
    "sns.lineplot(x=PC_values, y=pca.explained_variance_ratio_,  color='#42A7A9',marker=\"o\")   # marker=\"o\" shows markers as dots\n",
    "plt.title('Scree Plot',fontsize=20)\n",
    "plt.xlabel('Principal Component Number',fontsize=15)\n",
    "plt.ylabel('Variance Explained',fontsize=15)\n",
    "\n",
    "#plt.savefig( 'Figures/Scree_plot_protein_QC.png' , dpi=200,bbox_inches = 'tight' , pad_inches = 0.2 , facecolor='w')\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Individual residual variance method to identify outliers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Conduct PCA to identify outliers\n",
    "\n",
    "features = df_impute.columns[:]\n",
    "x = df_impute.loc[:, features].values\n",
    "\n",
    "pca = PCA(n_components=4) #determined using scree plot in the above cell\n",
    "principal_components = pca.fit_transform(x)\n",
    "\n",
    "\n",
    "\n",
    "#Calculate residuals to identify outliers\n",
    "#Formula is:     Residuals = Data matrix - (PCA scores * PCA loadings)  https://datascience.stackexchange.com/questions/29132/how-can-i-extract-the-residual-array-from-the-scikit-learn-pca-routine\n",
    "\n",
    "#Obtain PCA loadings (https://stackoverflow.com/questions/21217710/factor-loadings-using-sklearn)\n",
    "loadings = pca.components_.T * np.sqrt(pca.explained_variance_)  #Multiply each component by the square root of its corresponding eigenvalue\n",
    "\n",
    "residuals = df_impute - principal_components.dot(loadings.transpose())\n",
    "display(residuals)\n",
    "\n",
    "#Calculate individual residual variance by taking the Euclidean norm (L2 norm) of values for each sample\n",
    "#https://stackoverflow.com/questions/7741878/how-to-apply-numpy-linalg-norm-to-each-row-of-a-matrix\n",
    "l2_norm = pd.DataFrame(np.linalg.norm(residuals,axis=1),columns = ['Individual residual variance'], index = df.index)\n",
    "l2_norm_unordered = l2_norm.copy()\n",
    "l2_norm = l2_norm.sort_values(['Individual residual variance'], ascending=[False])\n",
    "display(l2_norm[:10])\n",
    "\n",
    "#Top 10 outliers\n",
    "outliers = list(l2_norm[:10].index.values)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Preparing data for bar graph \n",
    "l2_norm_unordered['Sample ID'] = l2_norm_unordered.index\n",
    "l2_norm_unordered = pd.concat([l2_norm_unordered, meta_data], axis = 1)\n",
    "l2_mean = l2_norm_unordered['Individual residual variance'].mean()\n",
    "l2_std = l2_norm_unordered['Individual residual variance'].std()\n",
    "\n",
    "\n",
    "#Bar graph showing individual residual variance with mean/median line\n",
    "g = sns.barplot(data=l2_norm_unordered,x=\"Sample ID\", y=\"Individual residual variance\",hue='Group',width=2)\n",
    "\n",
    "g2 = g.axes\n",
    "g2.set(xticklabels=[],xticks=[]) \n",
    "g2.axhline(l2_mean,color='r') #plot mean\n",
    "g2.axhline(l2_mean +(l2_std*3),color='r') #plot 3 standard deviations away from the mean\n",
    "g2.margins(x=0.01) #increase space between first and last bar and the edges \n",
    "\n",
    "#Add labels for outliers\n",
    "g2.text(3,610,\"1027114\")\n",
    "g2.text(36,510,\"1347882\")\n",
    "g2.text(215,430,\"Mean + 3sd\")\n",
    "g2.text(240,190,\"Mean\")  \n",
    "\n",
    "#plt.savefig( 'Figures/individual_residual_variance.png' , dpi=200,bbox_inches = 'tight' , pad_inches = 0.2 , facecolor='w')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_hist = df_impute.mean(axis = 0)\n",
    "sns.histplot(l2_norm, bins = 40,color='#BABAE2',edgecolor=\"k\",legend=False) \n",
    "\n",
    "plt.xlabel('Individual residual variance',fontsize=13)\n",
    "plt.ylabel('Count',fontsize=13) \n",
    "\n",
    "#plt.savefig( 'Figures/individual_residual_variance_distribution.png' , dpi=200,bbox_inches = 'tight' , pad_inches = 0.2 , facecolor='w')\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Counting number of outliers outside of mean plus/minus 3 standard deviations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Outlier detection step - Z-score method\n",
    "#The outliers should be removed before scaling\n",
    "df_norm = pd.DataFrame(StandardScaler().fit_transform(df_impute),columns=df_impute.columns, index=df_impute.index)\n",
    "\n",
    "df_norm = df_norm.to_numpy()\n",
    "outlier_count = np.sum((df_norm > 3) | (df_norm < -3),axis=1)\n",
    "outlier_df = pd.DataFrame(outlier_count, index = meta_data.index, columns = ['Outlier number'])\n",
    "outlier_df = outlier_df.sort_values(['Outlier number'], ascending=[False])\n",
    "\n",
    "\n",
    "\n",
    "#I originally determined the outliers by taking the outliers that are 3 standard deviations above the mean,\n",
    "#however if you look at the histogram below this is not accurate because the distribution is not a NORMAL DISTRIBUTION!\n",
    "#therefore it's better to just plot a histogram or barplot and determine by eye\n",
    "display(outlier_df[:10]) #Outliers 3 std above the mean (first 3)\n",
    "\n",
    "outlier_df['Outlier number'].mean()  #3.8108\n",
    "outlier_df['Outlier number'].std()  #12.17\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_hist = df_impute.mean(axis = 0)\n",
    "sns.histplot(outlier_df, bins = 50,color='#BABAE2',edgecolor=\"k\",legend=False) \n",
    "\n",
    "plt.title('Outlier count',fontsize=16)\n",
    "plt.xlabel('Number of outliers',fontsize=13)\n",
    "plt.ylabel('Count',fontsize=13) \n",
    "\n",
    "#plt.savefig( 'Figures/outlier_histogram.png' , dpi=200,bbox_inches = 'tight' , pad_inches = 0.2 , facecolor='w')\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Identifying outliers using pair plot method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Pair plots for first 4 principal components\n",
    "df2 = pd.DataFrame(data = principal_components, columns = ['PC1', 'PC2','PC3', 'PC4'], index = df.index)\n",
    "\n",
    "#Concatenate WHO information\n",
    "df3 = pd.concat([df2, meta_data], axis = 1)\n",
    "#display(df3)\n",
    "\n",
    "sns.pairplot(df3, hue='Group',height=1.8)   \n",
    "\n",
    "#plt.savefig( 'Figures/outlier_pairplot.png' , dpi=200,bbox_inches = 'tight' , pad_inches = 0.2 , facecolor='w')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Alternatively, can dentify outliers by looking at which points are at the extremes of the PC axes and checking with the PCA plot\n",
    "outliers_pc = df3.sort_values(['PC4'], ascending=[False])\n",
    "display(outliers_pc.iloc[:10,:])"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Visualising outliers with first 2 principal components"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#PCA for visualisation WITH FIRST 2 PRINCIPAL COMPONENTS AND OUTLIERS LABELLED\n",
    "features = df_impute.columns[:]\n",
    "x = df_impute.loc[:, features].values\n",
    "\n",
    "pca = PCA(n_components=2) #two principal components for data visualisation\n",
    "principal_components = pca.fit_transform(x)\n",
    "\n",
    "df2 = pd.DataFrame(data = principal_components, columns = ['PC1', 'PC2'], index = df.index)\n",
    "\n",
    "#Concatenate WHO information\n",
    "df3 = pd.concat([df2, meta_data], axis = 1)\n",
    "\n",
    "display(df3)\n",
    "\n",
    "#Subset outlier dataframe to plot outliers\n",
    "outlier_df = df3[df3.index.isin(outliers)]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "outlier_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sns.set_style(\"white\")\n",
    "\n",
    "sns.lmplot(\n",
    "    x='PC1', \n",
    "    y='PC2', \n",
    "    data=df3, \n",
    "    hue='WHO_status', \n",
    "    hue_order = ['0', '1-2', '3-4','5-7'],\n",
    "    fit_reg=False, \n",
    "    legend=False,\n",
    "    scatter_kws={\"s\": 20}\n",
    "    )\n",
    "#Note: I don't use the seaborn legend but check it matches with the seaborn legend\n",
    "\n",
    "for index,outlier in (enumerate(outliers)):\n",
    "    plt.text(df3.PC1[outlier],df3.PC2[outlier], outlier)\n",
    "\n",
    "\n",
    "plt.title('PCA for outlier detection',fontsize=20)\n",
    "plt.xlabel('PC1 (' + str(round(pca.explained_variance_ratio_[0]*100,2)) + '%)',fontsize=15)\n",
    "plt.ylabel('PC2 (' + str(round(pca.explained_variance_ratio_[1]*100,2)) + '%)',fontsize=15)\n",
    "plt.legend(framealpha=1, frameon = 'True', title=\"WHO status\",title_fontsize='large', prop={'size': 14}, bbox_to_anchor=(1.04, 0.7)) \n",
    "#This has more information on the bbox_to_anchor coordinates: https://stackoverflow.com/questions/4700614/how-to-put-the-legend-outside-the-plot\n",
    "\n",
    "#plt.savefig( 'Figures/PCA_protein_QC.png' , dpi=200,bbox_inches = 'tight' , pad_inches = 0.2 , facecolor='w')\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Removal of outliers prior to scaling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Remove outlier 1027114\n",
    "df_impute2 = df_impute.drop(1027114)\n",
    "meta_data = meta_data.drop(1027114)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Normalise by column i.e. each pathway is normalised across all patients\n",
    "#mean of the observed values becomes 0 and the standard deviation is 1\n",
    "\n",
    "\n",
    "df_norm = pd.DataFrame(StandardScaler().fit_transform(df_impute2),columns=df_impute2.columns, index=df_impute2.index)\n",
    "\n",
    "print(df_norm.max().max())\n",
    "print(df_norm.min().min())\n",
    "print(df_norm.mean(axis = 0)) #mean of 0  \n",
    "print(df_norm.std(axis = 0)) #sd of 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#One feature has a lower mean than the others (superoxide dismutase 2, mitochondrial)\n",
    "mean = pd.DataFrame(df_norm.mean(axis = 0))\n",
    "mean[[0]].idxmin()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_hist = df_norm.mean(axis = 0)\n",
    "sns.histplot(df_hist, bins = 50,color='#79C99E',edgecolor=\"k\") \n",
    "\n",
    "#The mean value for each metabolite has been plotted\n",
    "plt.title('Mean protein distribution',fontsize=16)\n",
    "plt.xlabel('Protein expression (e-15)',fontsize=13)\n",
    "plt.ylabel('Count',fontsize=13) ;\n",
    "\n",
    "#plt.savefig( 'Figures/mean_protein_distribution.png' , dpi=200,bbox_inches = 'tight' , pad_inches = 0.2 , facecolor='w')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "df_norm_np = df_norm.to_numpy()\n",
    "\n",
    "df_hist = df_norm_np.flatten()\n",
    "sns.histplot(df_hist, bins = 45,color='#79C99E',edgecolor=\"k\") \n",
    "\n",
    "#The mean value for each metabolite has been plotted\n",
    "plt.title('Protein distribution',fontsize=16)\n",
    "plt.xlabel('Protein expression',fontsize=13)\n",
    "plt.ylabel('Count',fontsize=13) ;\n",
    "\n",
    "#plt.savefig( 'Figures/protein_distribution.png' , dpi=200,bbox_inches = 'tight' , pad_inches = 0.2 , facecolor='w')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Carry out PCA\n",
    "\n",
    "features = df_norm.columns[:]\n",
    "x = df_norm.loc[:, features].values\n",
    "\n",
    "pca = PCA(n_components=2)\n",
    "principal_components = pca.fit_transform(x)\n",
    "df2 = pd.DataFrame(data = principal_components, columns = ['PC1', 'PC2'])\n",
    "\n",
    "#Restore original index\n",
    "df2 = df2.set_index(df_norm.index)\n",
    "\n",
    "#Concatenate WHO information\n",
    "df3 = pd.concat([df2, meta_data], axis = 1)\n",
    "\n",
    "#display(df3)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "sns.set_style(\"white\")\n",
    "\n",
    "\n",
    "sns.lmplot(\n",
    "    x='PC1', \n",
    "    y='PC2', \n",
    "    data=df3, \n",
    "    hue='WHO_status', \n",
    "    hue_order = ['0', '1-2', '3-4','5-7'],\n",
    "    fit_reg=False, #don't draw line of best fit\n",
    "    legend=False,\n",
    "    scatter_kws={\"s\": 20}\n",
    "    )\n",
    "#Note: I don't use the seaborn legend but check it matches with the seaborn legend\n",
    "\n",
    "\n",
    "plt.title('PCA for proteins',fontsize=20)\n",
    "plt.xlabel('PC1 (' + str(round(pca.explained_variance_ratio_[0]*100,2)) + '%)',fontsize=15,)\n",
    "plt.ylabel('PC2 (' + str(round(pca.explained_variance_ratio_[1]*100,2)) + '%)',fontsize=15)\n",
    "plt.legend(framealpha=1, frameon = 'True', title=\"WHO status\",title_fontsize='large', prop={'size': 14}, bbox_to_anchor=(1.04, 0.7)) \n",
    "\n",
    "#plt.savefig( 'Figures/proteins_PCA.png' , dpi=200,bbox_inches = 'tight' , pad_inches = 0.2 , facecolor='w')\n",
    "\n",
    "\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add metadata to the end of the df\n",
    "df_final = pd.concat([df_norm, meta_data],axis=1) \n",
    "#df_final.to_csv('Data/Su_COVID_proteomics_processed_names.csv')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert protein columns from names to UniProt IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Convert protein columns from names to Uniprot IDs \n",
    "\n",
    "#Read in pre-processed dataframe\n",
    "df = pd.read_csv('../Data/Su_COVID_proteomics_processed_names.csv', index_col=0)\n",
    "\n",
    "#Read in the conversion table\n",
    "conversion_tab = pd.read_csv('../Data/proteomics_metadata.csv', index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Separate dataframe into numerical data and meta data\n",
    "meta_data = df.iloc[:,-2:]\n",
    "df = df.iloc[:,:-2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Subset conversion table to only the proteins present in the data\n",
    "protein_names = list(df.columns)\n",
    "conversion_tab2 = conversion_tab[conversion_tab['gene_description'].isin(protein_names)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n"
     ]
    }
   ],
   "source": [
    "conversion_tab2 = conversion_tab2.sort_values(by='gene_description')\n",
    "\n",
    "#Check duplicates in the gene description column have the same Uniprot IDs (they do)\n",
    "conversion_tab2 = conversion_tab2.drop_duplicates(subset='uniprot')\n",
    "conversion_tab2 = conversion_tab2.reset_index(drop=True)\n",
    "\n",
    "from collections import Counter\n",
    "mylist = conversion_tab2.gene_description\n",
    "\n",
    "dups = [k for k,v in Counter(mylist).items() if v>1]\n",
    "print(dups)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_uniprot = df.rename(columns=dict(zip(conversion_tab2[\"gene_description\"], conversion_tab2[\"uniprot\"])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add metadata to the end of the df\n",
    "df_final = pd.concat([df_uniprot, meta_data],axis=1) \n",
    "#df_final.to_csv('Data/Su_COVID_proteomics_processed.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Imperial_Project2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
